{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Using ROBOKOP's Quick Question Answering Service"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The expand, enrich, and similarity services each offer bite-sized approaches to interacting with ROBOKOP.  The ROBOKOP \"quick\" service offers a slightly more complex approach to answering questions.   We'll be using the following function to call the quick service.  Note that calling the service requires posting a question.  We'll discuss the format of the question below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "robokop_server='robokop.renci.org'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def quick(question,max_results=None,output_format=None):\n",
    "    url=f'http://{robokop_server}:80/api/simple/quick/'\n",
    "    if max_results is not None:\n",
    "        url += f'?max_results={max_results}'\n",
    "    if output_format is not None:\n",
    "        j = '&' if '?' in url else '?'\n",
    "        url += f'{j}output_format={output_format}'\n",
    "    response = requests.post(url,json=question)\n",
    "    print( f\"Return Status: {response.status_code}\" )\n",
    "    if response.status_code == 200:\n",
    "        return response.json()\n",
    "    return response"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Question Format and Basic Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The question is a python dictionary.  It takes a key `machine_question`, which is a dictionary containing a list of `nodes` and a list of `edges`.   Each node object needs an integer `id`.   Any node can also have a `curie` specifying that node.  The edges define connections between the nodes using the identifiers of the nodes as source and targets. The following function shows how to construct a single-hop question.  It takes a specified node of `type1` and looks for any node of `type2`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def make_one_step_question(type1, id1, type2,rebuild = None):\n",
    "    question = {\n",
    "                'machine_question': {\n",
    "                    'nodes': [\n",
    "                        {\n",
    "                            'id': 'n0',\n",
    "                            'curie': id1,\n",
    "                            'type': type1\n",
    "                        },\n",
    "                        {\n",
    "                            'id': 'n1',\n",
    "                            'type': type2\n",
    "                        }\n",
    "                    ],\n",
    "                    'edges': [\n",
    "                        {\n",
    "                            'id': 'e0',\n",
    "                            'source_id': 'n0',\n",
    "                            'target_id': 'n1'\n",
    "                        }\n",
    "                    ]\n",
    "                }\n",
    "            }\n",
    "    if rebuild is not None and str(rebuild).upper() == 'TRUE':\n",
    "        question['rebuild'] = 'True'\n",
    "    return question"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Here, we will specify a one-hop question asking for the phenotypes associated with Fanconi Anemia (MONDO:0019391).  We first construct the question then use it to call the quick service.  In fact, this is how the expand function is implemented internally.  The output is returned in the KG standard answer format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "q = make_one_step_question('disease','MONDO:0019391','phenotypic_feature')\n",
    "r = quick(q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Uncomment to see a big long answer\n",
    "import json\n",
    "#print( json.dumps(r, indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The output can be simplified with a function like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def parse_answer(returnanswer):\n",
    "    #First, parse out the parts of the kg that we want, names and types\n",
    "    kg_node_names = { n['id']: n['name'] if 'name' in n else n['id'] for n in returnanswer['knowledge_graph']['nodes'] }\n",
    "    kg_edge_types = { e['id']: e['type'] for e in returnanswer['knowledge_graph']['edges']}\n",
    "    kg_edge_sources = { e['id']: e['edge_source'] for e in returnanswer['knowledge_graph']['edges']}\n",
    "    #now put the answers into a table, looking up node/edge info from the above dicts\n",
    "    #because of the question, we know that we're interested in the binding for \"n1\" and \"e0\"\n",
    "    #this is quick-n-dirty, so we're also limiting ourselves to the first edge, even if there is more than one\n",
    "    answers = [ {\"result_id\": answer['node_bindings']['n1'], \n",
    "                 \"result_name\": kg_node_names[answer['node_bindings']['n1']], #if 'name' in node else node['id'], \n",
    "                 \"type\":        kg_edge_types[answer['edge_bindings']['e0'][0]],\n",
    "                 \"source\":      kg_edge_sources[answer['edge_bindings']['e0'][0]],\n",
    "                 \"score\" :      answer['score']}\n",
    "              for answer in returnanswer['answers']]\n",
    "    return pd.DataFrame(answers)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>result_id</th>\n",
       "      <th>result_name</th>\n",
       "      <th>score</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HP:0005505</td>\n",
       "      <td>Refractory anemia</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HP:0100615</td>\n",
       "      <td>Ovarian neoplasm</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HP:0002488</td>\n",
       "      <td>Acute leukemia</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HP:0002860</td>\n",
       "      <td>Squamous cell carcinoma</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HP:0004430</td>\n",
       "      <td>Severe combined immunodeficiency</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HP:0002721</td>\n",
       "      <td>Immunodeficiency</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HP:0001402</td>\n",
       "      <td>Hepatocellular carcinoma</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HP:0000286</td>\n",
       "      <td>Epicanthus</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HP:0003002</td>\n",
       "      <td>Breast carcinoma</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HP:0004377</td>\n",
       "      <td>Hematological neoplasm</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HP:0007354</td>\n",
       "      <td>Amyotrophic lateral sclerosis</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HP:0008064</td>\n",
       "      <td>Ichthyosis</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HP:0006721</td>\n",
       "      <td>Acute lymphoblastic leukemia</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HP:0003254</td>\n",
       "      <td>Abnormality of DNA repair</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HP:0001915</td>\n",
       "      <td>Aplastic anemia</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HP:0000824</td>\n",
       "      <td>Growth hormone deficiency</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HP:0004808</td>\n",
       "      <td>Acute myeloid leukemia</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HP:0005560</td>\n",
       "      <td>Imbalanced hemoglobin synthesis</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HP:0100806</td>\n",
       "      <td>Sepsis</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HP:0005510</td>\n",
       "      <td>Transient erythroblastopenia</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HP:0001972</td>\n",
       "      <td>Macrocytic anemia</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HP:0001251</td>\n",
       "      <td>Ataxia</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HP:0011018</td>\n",
       "      <td>Abnormality of the cell cycle</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HP:0004431</td>\n",
       "      <td>Complement deficiency</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HP:0004810</td>\n",
       "      <td>Congenital hypoplastic anemia</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HP:0000789</td>\n",
       "      <td>Infertility</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HP:0000953</td>\n",
       "      <td>Hyperpigmentation of the skin</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HP:0001900</td>\n",
       "      <td>Increased hemoglobin</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HP:0001908</td>\n",
       "      <td>Hypoplastic anemia</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HP:0000135</td>\n",
       "      <td>Hypogonadism</td>\n",
       "      <td>1.619721</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>HP:0001053</td>\n",
       "      <td>Hypopigmented skin patches</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>HP:0001631</td>\n",
       "      <td>Atrial septal defect</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>HP:0002823</td>\n",
       "      <td>Abnormality of femur morphology</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>HP:0005344</td>\n",
       "      <td>Abnormal carotid artery morphology</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>HP:0004349</td>\n",
       "      <td>Reduced bone mineral density</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>HP:0000218</td>\n",
       "      <td>High palate</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>HP:0000486</td>\n",
       "      <td>Strabismus</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>HP:0008053</td>\n",
       "      <td>Aplasia/Hypoplasia of the iris</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>HP:0000813</td>\n",
       "      <td>Bicornuate uterus</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>HP:0001639</td>\n",
       "      <td>Hypertrophic cardiomyopathy</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>HP:0100026</td>\n",
       "      <td>Arteriovenous malformation</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>HP:0000864</td>\n",
       "      <td>Abnormality of the hypothalamus-pituitary axis</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>HP:0007565</td>\n",
       "      <td>Multiple cafe-au-lait spots</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>HP:0008572</td>\n",
       "      <td>External ear malformation</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>HP:0002007</td>\n",
       "      <td>Frontal bossing</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>HP:0004209</td>\n",
       "      <td>Clinodactyly of the 5th finger</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>HP:0000957</td>\n",
       "      <td>Cafe-au-lait spot</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>HP:0002023</td>\n",
       "      <td>Anal atresia</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>HP:0001903</td>\n",
       "      <td>Anemia</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>HP:0100760</td>\n",
       "      <td>Clubbing of toes</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>HP:0000639</td>\n",
       "      <td>Nystagmus</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>HP:0002827</td>\n",
       "      <td>Hip dislocation</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>HP:0001199</td>\n",
       "      <td>Triphalangeal thumb</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>HP:0012745</td>\n",
       "      <td>Short palpebral fissure</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>HP:0006101</td>\n",
       "      <td>Finger syndactyly</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>HP:0003022</td>\n",
       "      <td>Hypoplasia of the ulna</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>HP:0002251</td>\n",
       "      <td>Aganglionic megacolon</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>HP:0001636</td>\n",
       "      <td>Tetralogy of Fallot</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>HP:0000483</td>\n",
       "      <td>Astigmatism</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>HP:0006501</td>\n",
       "      <td>Aplasia/Hypoplasia of the radius</td>\n",
       "      <td>0.517653</td>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>250 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      result_id                                     result_name     score  \\\n",
       "0    HP:0005505                               Refractory anemia  1.619721   \n",
       "1    HP:0100615                                Ovarian neoplasm  1.619721   \n",
       "2    HP:0002488                                  Acute leukemia  1.619721   \n",
       "3    HP:0002860                         Squamous cell carcinoma  1.619721   \n",
       "4    HP:0004430                Severe combined immunodeficiency  1.619721   \n",
       "5    HP:0002721                                Immunodeficiency  1.619721   \n",
       "6    HP:0001402                        Hepatocellular carcinoma  1.619721   \n",
       "7    HP:0000286                                      Epicanthus  1.619721   \n",
       "8    HP:0003002                                Breast carcinoma  1.619721   \n",
       "9    HP:0004377                          Hematological neoplasm  1.619721   \n",
       "10   HP:0007354                   Amyotrophic lateral sclerosis  1.619721   \n",
       "11   HP:0008064                                      Ichthyosis  1.619721   \n",
       "12   HP:0006721                    Acute lymphoblastic leukemia  1.619721   \n",
       "13   HP:0003254                       Abnormality of DNA repair  1.619721   \n",
       "14   HP:0001915                                 Aplastic anemia  1.619721   \n",
       "15   HP:0000824                       Growth hormone deficiency  1.619721   \n",
       "16   HP:0004808                          Acute myeloid leukemia  1.619721   \n",
       "17   HP:0005560                 Imbalanced hemoglobin synthesis  1.619721   \n",
       "18   HP:0100806                                          Sepsis  1.619721   \n",
       "19   HP:0005510                    Transient erythroblastopenia  1.619721   \n",
       "20   HP:0001972                               Macrocytic anemia  1.619721   \n",
       "21   HP:0001251                                          Ataxia  1.619721   \n",
       "22   HP:0011018                   Abnormality of the cell cycle  1.619721   \n",
       "23   HP:0004431                           Complement deficiency  1.619721   \n",
       "24   HP:0004810                   Congenital hypoplastic anemia  1.619721   \n",
       "25   HP:0000789                                     Infertility  1.619721   \n",
       "26   HP:0000953                   Hyperpigmentation of the skin  1.619721   \n",
       "27   HP:0001900                            Increased hemoglobin  1.619721   \n",
       "28   HP:0001908                              Hypoplastic anemia  1.619721   \n",
       "29   HP:0000135                                    Hypogonadism  1.619721   \n",
       "..          ...                                             ...       ...   \n",
       "220  HP:0001053                      Hypopigmented skin patches  0.517653   \n",
       "221  HP:0001631                            Atrial septal defect  0.517653   \n",
       "222  HP:0002823                 Abnormality of femur morphology  0.517653   \n",
       "223  HP:0005344              Abnormal carotid artery morphology  0.517653   \n",
       "224  HP:0004349                    Reduced bone mineral density  0.517653   \n",
       "225  HP:0000218                                     High palate  0.517653   \n",
       "226  HP:0000486                                      Strabismus  0.517653   \n",
       "227  HP:0008053                  Aplasia/Hypoplasia of the iris  0.517653   \n",
       "228  HP:0000813                               Bicornuate uterus  0.517653   \n",
       "229  HP:0001639                     Hypertrophic cardiomyopathy  0.517653   \n",
       "230  HP:0100026                      Arteriovenous malformation  0.517653   \n",
       "231  HP:0000864  Abnormality of the hypothalamus-pituitary axis  0.517653   \n",
       "232  HP:0007565                     Multiple cafe-au-lait spots  0.517653   \n",
       "233  HP:0008572                       External ear malformation  0.517653   \n",
       "234  HP:0002007                                 Frontal bossing  0.517653   \n",
       "235  HP:0004209                  Clinodactyly of the 5th finger  0.517653   \n",
       "236  HP:0000957                               Cafe-au-lait spot  0.517653   \n",
       "237  HP:0002023                                    Anal atresia  0.517653   \n",
       "238  HP:0001903                                          Anemia  0.517653   \n",
       "239  HP:0100760                                Clubbing of toes  0.517653   \n",
       "240  HP:0000639                                       Nystagmus  0.517653   \n",
       "241  HP:0002827                                 Hip dislocation  0.517653   \n",
       "242  HP:0001199                             Triphalangeal thumb  0.517653   \n",
       "243  HP:0012745                         Short palpebral fissure  0.517653   \n",
       "244  HP:0006101                               Finger syndactyly  0.517653   \n",
       "245  HP:0003022                          Hypoplasia of the ulna  0.517653   \n",
       "246  HP:0002251                           Aganglionic megacolon  0.517653   \n",
       "247  HP:0001636                             Tetralogy of Fallot  0.517653   \n",
       "248  HP:0000483                                     Astigmatism  0.517653   \n",
       "249  HP:0006501                Aplasia/Hypoplasia of the radius  0.517653   \n",
       "\n",
       "                            source           type  \n",
       "0    biolink.disease_get_phenotype  has_phenotype  \n",
       "1    biolink.disease_get_phenotype  has_phenotype  \n",
       "2    biolink.disease_get_phenotype  has_phenotype  \n",
       "3    biolink.disease_get_phenotype  has_phenotype  \n",
       "4    biolink.disease_get_phenotype  has_phenotype  \n",
       "5    biolink.disease_get_phenotype  has_phenotype  \n",
       "6    biolink.disease_get_phenotype  has_phenotype  \n",
       "7    biolink.disease_get_phenotype  has_phenotype  \n",
       "8    biolink.disease_get_phenotype  has_phenotype  \n",
       "9    biolink.disease_get_phenotype  has_phenotype  \n",
       "10   biolink.disease_get_phenotype  has_phenotype  \n",
       "11   biolink.disease_get_phenotype  has_phenotype  \n",
       "12   biolink.disease_get_phenotype  has_phenotype  \n",
       "13   biolink.disease_get_phenotype  has_phenotype  \n",
       "14   biolink.disease_get_phenotype  has_phenotype  \n",
       "15   biolink.disease_get_phenotype  has_phenotype  \n",
       "16   biolink.disease_get_phenotype  has_phenotype  \n",
       "17   biolink.disease_get_phenotype  has_phenotype  \n",
       "18   biolink.disease_get_phenotype  has_phenotype  \n",
       "19   biolink.disease_get_phenotype  has_phenotype  \n",
       "20   biolink.disease_get_phenotype  has_phenotype  \n",
       "21   biolink.disease_get_phenotype  has_phenotype  \n",
       "22   biolink.disease_get_phenotype  has_phenotype  \n",
       "23   biolink.disease_get_phenotype  has_phenotype  \n",
       "24   biolink.disease_get_phenotype  has_phenotype  \n",
       "25   biolink.disease_get_phenotype  has_phenotype  \n",
       "26   biolink.disease_get_phenotype  has_phenotype  \n",
       "27   biolink.disease_get_phenotype  has_phenotype  \n",
       "28   biolink.disease_get_phenotype  has_phenotype  \n",
       "29   biolink.disease_get_phenotype  has_phenotype  \n",
       "..                             ...            ...  \n",
       "220  biolink.disease_get_phenotype  has_phenotype  \n",
       "221  biolink.disease_get_phenotype  has_phenotype  \n",
       "222  biolink.disease_get_phenotype  has_phenotype  \n",
       "223  biolink.disease_get_phenotype  has_phenotype  \n",
       "224  biolink.disease_get_phenotype  has_phenotype  \n",
       "225  biolink.disease_get_phenotype  has_phenotype  \n",
       "226  biolink.disease_get_phenotype  has_phenotype  \n",
       "227  biolink.disease_get_phenotype  has_phenotype  \n",
       "228  biolink.disease_get_phenotype  has_phenotype  \n",
       "229  biolink.disease_get_phenotype  has_phenotype  \n",
       "230  biolink.disease_get_phenotype  has_phenotype  \n",
       "231  biolink.disease_get_phenotype  has_phenotype  \n",
       "232  biolink.disease_get_phenotype  has_phenotype  \n",
       "233  biolink.disease_get_phenotype  has_phenotype  \n",
       "234  biolink.disease_get_phenotype  has_phenotype  \n",
       "235  biolink.disease_get_phenotype  has_phenotype  \n",
       "236  biolink.disease_get_phenotype  has_phenotype  \n",
       "237  biolink.disease_get_phenotype  has_phenotype  \n",
       "238  biolink.disease_get_phenotype  has_phenotype  \n",
       "239  biolink.disease_get_phenotype  has_phenotype  \n",
       "240  biolink.disease_get_phenotype  has_phenotype  \n",
       "241  biolink.disease_get_phenotype  has_phenotype  \n",
       "242  biolink.disease_get_phenotype  has_phenotype  \n",
       "243  biolink.disease_get_phenotype  has_phenotype  \n",
       "244  biolink.disease_get_phenotype  has_phenotype  \n",
       "245  biolink.disease_get_phenotype  has_phenotype  \n",
       "246  biolink.disease_get_phenotype  has_phenotype  \n",
       "247  biolink.disease_get_phenotype  has_phenotype  \n",
       "248  biolink.disease_get_phenotype  has_phenotype  \n",
       "249  biolink.disease_get_phenotype  has_phenotype  \n",
       "\n",
       "[250 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_answer(r)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Multi-step questions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "It's straightforward to generalize this linear query to an N-item path.  This function constructs a question from a list of types and a list of identifiers.  The types are the types of nodes that will be traversed along the path, and the identifiers represent fixed elements in the path.  The length of types and ids should be equal, and free nodes should specify an id of None:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def make_N_step_question(types,curies,rebuild = None):\n",
    "    question = {\n",
    "                'machine_question': {\n",
    "                    'nodes': [],\n",
    "                    'edges': []\n",
    "                }\n",
    "            }\n",
    "    if rebuild is not None and str(rebuild).upper() == 'TRUE':\n",
    "        question['rebuild'] = 'True'\n",
    "    ei = 0\n",
    "    for i,t in enumerate(types):\n",
    "        newnode = {'id': f'n{i}', 'type': t}\n",
    "        if curies[i] is not None:\n",
    "            newnode['curie'] = curies[i]\n",
    "        question['machine_question']['nodes'].append(newnode)\n",
    "        if i > 0:\n",
    "            question['machine_question']['edges'].append( {'id': f'e{ei}','source_id': f'n{i-1}', 'target_id': f'n{i}'})\n",
    "            ei += 1\n",
    "    return question"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "We can recapitulate our previous question with this new function like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "newq = make_N_step_question(['disease','phenotypic_feature'],['MONDO:0019391',None])\n",
    "q == newq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Now we could expand to a longer query.  The following question will start at the disease MONDO:0019391, go to a gene, and from there to a biological process or activity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "two_step_question = make_N_step_question(['disease','gene','biological_process_or_activity'],['MONDO:0019391',None,None])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "two_step_answer = quick(two_step_question)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "We can extract the node names along the paths with a simple function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def extract_node_names(returnanswer):\n",
    "    #make a dict from identifiers (like MONDO:0019391) to names (like \"Fanconi Anemia\")\n",
    "    kg_node_names = { n['id']: n['name'] if 'name' in n else n['id'] for n in returnanswer['knowledge_graph']['nodes'] }\n",
    "    #For every answer, make a dict from the query node ids (like n0) to the node name (like Fanconia Anemia)\n",
    "    nodes = [\n",
    "                {q_node: kg_node_names[ q_id ] for q_node, q_id in answer['node_bindings'].items()}\n",
    "                for answer in returnanswer['answers']\n",
    "            ]\n",
    "    #Turn this list of dicts into a table\n",
    "    return pd.DataFrame(nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n0</th>\n",
       "      <th>n1</th>\n",
       "      <th>n2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>BRIP1</td>\n",
       "      <td>response to toxic substance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FANCG</td>\n",
       "      <td>response to radiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51C</td>\n",
       "      <td>DNA recombination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FANCA</td>\n",
       "      <td>regulation of cell proliferation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FANCA</td>\n",
       "      <td>female gonad development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FANCL</td>\n",
       "      <td>regulation of cell proliferation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>female gonad development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>inner cell mass cell proliferation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>hemopoiesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>PALB2</td>\n",
       "      <td>inner cell mass cell proliferation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>PALB2</td>\n",
       "      <td>animal organ morphogenesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>DNA recombination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>response to UV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MAD2L2</td>\n",
       "      <td>positive regulation of isotype switching</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>endodeoxyribonuclease activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MAD2L2</td>\n",
       "      <td>cell division</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FANCF</td>\n",
       "      <td>protein ubiquitination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MX1</td>\n",
       "      <td>membrane fusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>endothelial cell apoptotic process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>positive regulation of cell adhesion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>microglial cell activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MX1</td>\n",
       "      <td>defense response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>inflammatory response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>astrocyte activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>osteoclast differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>extracellular matrix organization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>humoral immune response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MX1</td>\n",
       "      <td>innate immune response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>epithelial cell proliferation involved in sali...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>leukocyte tethering or rolling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>identical protein binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>transcription-coupled nucleotide-excision repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>single-stranded DNA endodeoxyribonuclease acti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>cellular response to UV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MAD2L2</td>\n",
       "      <td>RNA polymerase II activating transcription fac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>double-stranded DNA binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>enzyme binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MAD2L2</td>\n",
       "      <td>negative regulation of protein catabolic process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>mitotic recombination-dependent replication fo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>single-stranded DNA-dependent ATPase activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>nucleotide-excision repair involved in interst...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>DNA binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>DNA unwinding involved in DNA replication</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MAD2L2</td>\n",
       "      <td>negative regulation of canonical Wnt signaling...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MAD2L2</td>\n",
       "      <td>positive regulation of double-strand break rep...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>protein N-terminus binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>chromatin binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>telomerase inhibitor activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>3' overhang single-stranded DNA endodeoxyribon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>negative regulation of telomere maintenance vi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>protein homooligomerization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>TFIID-class transcription factor complex binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MAD2L2</td>\n",
       "      <td>regulation of cell growth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>ATP binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>damaged DNA binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>cellular response to DNA damage stimulus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>strand invasion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>recombinase activity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>250 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 n0      n1                                                 n2\n",
       "0    Fanconi anemia   BRIP1                        response to toxic substance\n",
       "1    Fanconi anemia   FANCG                              response to radiation\n",
       "2    Fanconi anemia  RAD51C                                  DNA recombination\n",
       "3    Fanconi anemia   FANCA                   regulation of cell proliferation\n",
       "4    Fanconi anemia   FANCA                           female gonad development\n",
       "5    Fanconi anemia   FANCL                   regulation of cell proliferation\n",
       "6    Fanconi anemia   BRCA2                           female gonad development\n",
       "7    Fanconi anemia   BRCA2                 inner cell mass cell proliferation\n",
       "8    Fanconi anemia   BRCA2                                        hemopoiesis\n",
       "9    Fanconi anemia   PALB2                 inner cell mass cell proliferation\n",
       "10   Fanconi anemia   PALB2                         animal organ morphogenesis\n",
       "11   Fanconi anemia   RAD51                                  DNA recombination\n",
       "12   Fanconi anemia   ERCC4                                     response to UV\n",
       "13   Fanconi anemia  MAD2L2           positive regulation of isotype switching\n",
       "14   Fanconi anemia   ERCC4                     endodeoxyribonuclease activity\n",
       "15   Fanconi anemia  MAD2L2                                      cell division\n",
       "16   Fanconi anemia   FANCF                             protein ubiquitination\n",
       "17   Fanconi anemia     MX1                                    membrane fusion\n",
       "18   Fanconi anemia     TNF                 endothelial cell apoptotic process\n",
       "19   Fanconi anemia     TNF               positive regulation of cell adhesion\n",
       "20   Fanconi anemia     TNF                         microglial cell activation\n",
       "21   Fanconi anemia     MX1                                   defense response\n",
       "22   Fanconi anemia     TNF                              inflammatory response\n",
       "23   Fanconi anemia     TNF                               astrocyte activation\n",
       "24   Fanconi anemia     TNF                         osteoclast differentiation\n",
       "25   Fanconi anemia     TNF                  extracellular matrix organization\n",
       "26   Fanconi anemia     TNF                            humoral immune response\n",
       "27   Fanconi anemia     MX1                             innate immune response\n",
       "28   Fanconi anemia     TNF  epithelial cell proliferation involved in sali...\n",
       "29   Fanconi anemia     TNF                     leukocyte tethering or rolling\n",
       "..              ...     ...                                                ...\n",
       "220  Fanconi anemia   RAD51                          identical protein binding\n",
       "221  Fanconi anemia   ERCC4   transcription-coupled nucleotide-excision repair\n",
       "222  Fanconi anemia   ERCC4  single-stranded DNA endodeoxyribonuclease acti...\n",
       "223  Fanconi anemia   ERCC4                            cellular response to UV\n",
       "224  Fanconi anemia  MAD2L2  RNA polymerase II activating transcription fac...\n",
       "225  Fanconi anemia   RAD51                        double-stranded DNA binding\n",
       "226  Fanconi anemia   RAD51                                     enzyme binding\n",
       "227  Fanconi anemia  MAD2L2   negative regulation of protein catabolic process\n",
       "228  Fanconi anemia   RAD51  mitotic recombination-dependent replication fo...\n",
       "229  Fanconi anemia   RAD51      single-stranded DNA-dependent ATPase activity\n",
       "230  Fanconi anemia   ERCC4                                         DNA repair\n",
       "231  Fanconi anemia   ERCC4  nucleotide-excision repair involved in interst...\n",
       "232  Fanconi anemia   ERCC4                                        DNA binding\n",
       "233  Fanconi anemia   RAD51          DNA unwinding involved in DNA replication\n",
       "234  Fanconi anemia  MAD2L2  negative regulation of canonical Wnt signaling...\n",
       "235  Fanconi anemia  MAD2L2  positive regulation of double-strand break rep...\n",
       "236  Fanconi anemia   ERCC4                         protein N-terminus binding\n",
       "237  Fanconi anemia   RAD51                                  chromatin binding\n",
       "238  Fanconi anemia   ERCC4                      telomerase inhibitor activity\n",
       "239  Fanconi anemia   ERCC4  3' overhang single-stranded DNA endodeoxyribon...\n",
       "240  Fanconi anemia   ERCC4  negative regulation of telomere maintenance vi...\n",
       "241  Fanconi anemia   RAD51                        protein homooligomerization\n",
       "242  Fanconi anemia   ERCC4   TFIID-class transcription factor complex binding\n",
       "243  Fanconi anemia  MAD2L2                          regulation of cell growth\n",
       "244  Fanconi anemia   RAD51                                        ATP binding\n",
       "245  Fanconi anemia   ERCC4                                damaged DNA binding\n",
       "246  Fanconi anemia   RAD51           cellular response to DNA damage stimulus\n",
       "247  Fanconi anemia   ERCC4                      interstrand cross-link repair\n",
       "248  Fanconi anemia   RAD51                                    strand invasion\n",
       "249  Fanconi anemia   RAD51                               recombinase activity\n",
       "\n",
       "[250 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "two_step_nodes = extract_node_names(two_step_answer)\n",
    "two_step_nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output Format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Version 0.9, output_format=message"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results above are returned in v0.9 of the Translator standard for graph answers.  In this format, one graph is returned, containing all of the nodes and edges required for any of the answers.  Then the list of answers is given, binding nodes and edges in the question definition to nodes and edges in the shared knowledge graph.  This allows greater efficiency and specificity in relating outputs to question parameters.\n",
    "\n",
    "The response has 3 keys returning\n",
    "1. The json formatted question that was asked \n",
    "2. The shared knowledge graph \n",
    "3. The answers as defined on that graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['question_graph', 'knowledge_graph', 'answers'])\n"
     ]
    }
   ],
   "source": [
    "print(two_step_answer.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The knowledge graph consists of nodes (containing all the information about the entity) and edges (containing information about the relationship between entities)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['nodes', 'edges'])\n"
     ]
    }
   ],
   "source": [
    "print(two_step_answer['knowledge_graph'].keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's what the first 2 nodes and the first edge look like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "    {\n",
      "        \"congenital abnormality\": true,\n",
      "        \"name\": \"Fanconi anemia\",\n",
      "        \"nutritional or metabolic disease\": true,\n",
      "        \"id\": \"MONDO:0019391\",\n",
      "        \"equivalent_identifiers\": [\n",
      "            \"MESH:D005199\",\n",
      "            \"MEDDRA:10016218\",\n",
      "            \"MONDO:0019391\",\n",
      "            \"MEDDRA:10055206\",\n",
      "            \"UMLS:C0015625\",\n",
      "            \"ORPHANET:84\",\n",
      "            \"DOID:13636\"\n",
      "        ],\n",
      "        \"type\": \"disease\",\n",
      "        \"rare disease\": true,\n",
      "        \"syndromic disease\": true\n",
      "    },\n",
      "    {\n",
      "        \"name\": \"UBE2T\",\n",
      "        \"location\": \"1q32.1\",\n",
      "        \"locus_group\": \"protein-coding gene\",\n",
      "        \"gene_family\": [\n",
      "            \"Ubiquitin conjugating enzymes E2\",\n",
      "            \"FA complementation groups\"\n",
      "        ],\n",
      "        \"chromosome\": \"1\",\n",
      "        \"id\": \"HGNC:25009\",\n",
      "        \"equivalent_identifiers\": [\n",
      "            \"UniProtKB:Q9NPD8\",\n",
      "            \"NCBIGene:29089\",\n",
      "            \"HGNC:25009\",\n",
      "            \"ENSEMBL:ENSG00000077152\",\n",
      "            \"HGNC.SYMBOL:UBE2T\",\n",
      "            \"UniProtKB:Q9NPD8\"\n",
      "        ],\n",
      "        \"taxon\": \"9606\",\n",
      "        \"gene_family_id\": [\n",
      "            102,\n",
      "            548\n",
      "        ],\n",
      "        \"type\": \"gene\"\n",
      "    }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(two_step_answer['knowledge_graph']['nodes'][:2],indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"relation\": \"PHAROS:gene_involved\",\n",
      "    \"target_id\": \"HGNC:25009\",\n",
      "    \"edge_source\": \"pharos.disease_get_gene\",\n",
      "    \"publications\": [],\n",
      "    \"id\": \"4257632\",\n",
      "    \"predicate_id\": \"NCIT:R176\",\n",
      "    \"source_database\": \"pharos\",\n",
      "    \"source_id\": \"MONDO:0019391\",\n",
      "    \"type\": \"disease_to_gene_association\",\n",
      "    \"ctime\": 1543738834.740142,\n",
      "    \"relation_label\": \"gene_involved\",\n",
      "    \"weight\": 0.25832639515782074\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(two_step_answer['knowledge_graph']['edges'][1],indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's what the first answer looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"node_bindings\": {\n",
      "        \"n0\": \"MONDO:0019391\",\n",
      "        \"n1\": \"HGNC:20473\",\n",
      "        \"n2\": \"GO:0009636\"\n",
      "    },\n",
      "    \"edge_bindings\": {\n",
      "        \"e0\": [\n",
      "            \"4257531\",\n",
      "            \"6938566\",\n",
      "            \"6938600\"\n",
      "        ],\n",
      "        \"e1\": [\n",
      "            \"4294578\",\n",
      "            \"1094816\"\n",
      "        ]\n",
      "    },\n",
      "    \"score\": 0.3959587237470695\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(two_step_answer['answers'][0],indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The answers are ranked by score, so this is the highest ranking answer, and the score is given.  The `node_bindings` relate the identifiers of the questions (`n0`, `n1`, `n2`) to the identifiers in the knowledge graph (`MONDO:0019391`, `HGNC:20473`, `GO:0009636`) and we could use these identifiers to look up further information about those nodes in the graph.\n",
    "\n",
    "Similarly, the `e0` edge (connecting `n0` and `n1` in the question), is mapped to several possible edges in the graph, and we could look up the edge information using the identifier.  For instance, we can find the first edge between for `e0`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.3959587237470695,\n",
       " 0.39562714985581743,\n",
       " 0.3930911628130927,\n",
       " 0.38680121546632074,\n",
       " 0.38680121546632074,\n",
       " 0.38237448152146924,\n",
       " 0.37742069307012543,\n",
       " 0.37742069307012543,\n",
       " 0.37742069307012543,\n",
       " 0.35967233858255104,\n",
       " 0.35967233858255104,\n",
       " 0.3328828490620355,\n",
       " 0.3328828490620355,\n",
       " 0.3328828490620355,\n",
       " 0.3328828490620355,\n",
       " 0.3328828490620355,\n",
       " 0.330896351413593,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.22921193146495913,\n",
       " 0.22921193146495913,\n",
       " 0.22921193146495913,\n",
       " 0.22921193146495913,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.2252964653481522,\n",
       " 0.2252964653481522,\n",
       " 0.2252964653481522,\n",
       " 0.2252964653481522,\n",
       " 0.2252964653481522,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.21598706393112438,\n",
       " 0.21598706393112438,\n",
       " 0.21598706393112438,\n",
       " 0.21598706393112438,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scores = [a['score'] for a in two_step_answer['answers']]\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"relation\": \"PHAROS:gene_involved\",\n",
      "    \"target_id\": \"HGNC:20473\",\n",
      "    \"edge_source\": \"pharos.disease_get_gene\",\n",
      "    \"publications\": [],\n",
      "    \"id\": \"4257531\",\n",
      "    \"predicate_id\": \"NCIT:R176\",\n",
      "    \"source_database\": \"pharos\",\n",
      "    \"source_id\": \"MONDO:0019391\",\n",
      "    \"type\": \"disease_to_gene_association\",\n",
      "    \"ctime\": 1543738834.73799,\n",
      "    \"relation_label\": \"gene_involved\",\n",
      "    \"weight\": 0.25832639515782074\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "my_edge = list(filter(lambda x: x['id'] == \"4257531\", two_step_answer['knowledge_graph']['edges']))[0]\n",
    "print(json.dumps(my_edge,indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### v0.8, output_format=dense"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For backwards compatibility, output can also be returned in v0.8 of the KG format.  To return output in this format, pass the argument `output_format=dense` to the url.   The `output_format` defaults to `message` which specifies v0.9."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "two_step_answer_dense = quick(two_step_question,output_format='dense')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['datetime', 'id', 'message', 'response_code', 'result_list'])"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "two_step_answer_dense.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'confidence': 0.3959587237470695,\n",
       " 'id': '5939821d-f96e-442c-ab37-059ffe1087e4',\n",
       " 'result_graph': {'edge_list': [{'confidence': 0.25832639515782074,\n",
       "    'num_publications': None,\n",
       "    'provided_by': 'pharos.disease_get_gene',\n",
       "    'publications': '',\n",
       "    'source_id': 'MONDO:0019391',\n",
       "    'target_id': 'HGNC:20473',\n",
       "    'type': 'disease_to_gene_association'},\n",
       "   {'confidence': 0.4071474314830641,\n",
       "    'num_publications': None,\n",
       "    'provided_by': 'biolink.disease_get_gene',\n",
       "    'publications': 'PMID:20639400',\n",
       "    'source_id': 'MONDO:0019391',\n",
       "    'target_id': 'HGNC:20473',\n",
       "    'type': 'biomarker_for'},\n",
       "   {'confidence': 0.9998780282217139,\n",
       "    'num_publications': None,\n",
       "    'provided_by': 'biolink.disease_get_gene',\n",
       "    'publications': 'PMID:25980754,PMID:19127258,PMID:16116423,PMID:16153896,PMID:16116424,PMID:27498913,PMID:22006311,PMID:24556621,PMID:26689913,PMID:26681312,PMID:17033622,PMID:26720728,PMID:25186627,PMID:25452441,PMID:26921362,PMID:21964575,PMID:26786923,PMID:21127055,PMID:24763289,PMID:26845104,PMID:27153395,PMID:18628483,PMID:20639400,PMID:26315354,PMID:22792074,PMID:20159562',\n",
       "    'source_id': 'MONDO:0019391',\n",
       "    'target_id': 'HGNC:20473',\n",
       "    'type': 'gene_associated_with_condition'},\n",
       "   {'confidence': 0.25832639515782074,\n",
       "    'num_publications': None,\n",
       "    'provided_by': 'biolink.gene_get_process_or_function',\n",
       "    'publications': '',\n",
       "    'source_id': 'HGNC:20473',\n",
       "    'target_id': 'GO:0009636',\n",
       "    'type': 'actively_involved_in'},\n",
       "   {'confidence': 0.25832639515782074,\n",
       "    'num_publications': None,\n",
       "    'provided_by': 'quickgo.go_term_to_gene_annotation',\n",
       "    'publications': '',\n",
       "    'source_id': 'HGNC:20473',\n",
       "    'target_id': 'GO:0009636',\n",
       "    'type': 'actively_involved_in'}],\n",
       "  'node_list': [{'description': 'Fanconi anemia',\n",
       "    'id': 'MONDO:0019391',\n",
       "    'name': 'Fanconi anemia',\n",
       "    'type': 'disease'},\n",
       "   {'description': 'BRIP1',\n",
       "    'id': 'HGNC:20473',\n",
       "    'name': 'BRIP1',\n",
       "    'type': 'gene'},\n",
       "   {'description': 'response to toxic substance',\n",
       "    'id': 'GO:0009636',\n",
       "    'name': 'response to toxic substance',\n",
       "    'type': 'biological_process_or_activity'}]},\n",
       " 'result_type': 'individual query answer',\n",
       " 'text': 'response to toxic substance'}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "two_step_answer_dense['result_list'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Specifying multiple fixed nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The examples above have all started from one known node and expanded from it, sometimes in multiple steps.  It's also possible to have more than one node specified.  For instance, if we wanted to look for genes that link Fanconi Aneima (MONDO:0019391) and DNA repair (GO:0006281), we could rerun our previous query, but setting the final curie, like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "two_step_question_fixed_ends = \\\n",
    "   make_N_step_question(['disease','gene','biological_process_or_activity'],['MONDO:0019391',None,'GO:0006281'])\n",
    "two_step_answer_fixed_ends = quick(two_step_question_fixed_ends)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n0</th>\n",
       "      <th>n1</th>\n",
       "      <th>n2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FANCC</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FANCG</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51C</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FANCA</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FANCL</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>SLX4</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>XRCC2</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>UBE2T</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FANCD2</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>CENPS</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>XPA</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>WRN</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>FAN1</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51D</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>SLX1A</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>SLX1B</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>BLM</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>DDX11</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>RAD51AP1</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>MUS81</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Fanconi anemia</td>\n",
       "      <td>BOD1L1</td>\n",
       "      <td>DNA repair</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                n0        n1          n2\n",
       "0   Fanconi anemia     FANCC  DNA repair\n",
       "1   Fanconi anemia     FANCG  DNA repair\n",
       "2   Fanconi anemia    RAD51C  DNA repair\n",
       "3   Fanconi anemia     FANCA  DNA repair\n",
       "4   Fanconi anemia     FANCL  DNA repair\n",
       "5   Fanconi anemia      SLX4  DNA repair\n",
       "6   Fanconi anemia     RAD51  DNA repair\n",
       "7   Fanconi anemia     ERCC4  DNA repair\n",
       "8   Fanconi anemia     XRCC2  DNA repair\n",
       "9   Fanconi anemia     UBE2T  DNA repair\n",
       "10  Fanconi anemia    FANCD2  DNA repair\n",
       "11  Fanconi anemia     CENPS  DNA repair\n",
       "12  Fanconi anemia       XPA  DNA repair\n",
       "13  Fanconi anemia       WRN  DNA repair\n",
       "14  Fanconi anemia      FAN1  DNA repair\n",
       "15  Fanconi anemia    RAD51D  DNA repair\n",
       "16  Fanconi anemia     SLX1A  DNA repair\n",
       "17  Fanconi anemia     SLX1B  DNA repair\n",
       "18  Fanconi anemia       BLM  DNA repair\n",
       "19  Fanconi anemia     DDX11  DNA repair\n",
       "20  Fanconi anemia  RAD51AP1  DNA repair\n",
       "21  Fanconi anemia     MUS81  DNA repair\n",
       "22  Fanconi anemia    BOD1L1  DNA repair"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extract_node_names(two_step_answer_fixed_ends)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Non-path queries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "So far, we've only looked at linear paths.  But the question format is actually more general than that - we can define a path pattern generally. So for instance, in the above query, we find all genes that are linked to both FA and DNA repair.  But what if we wanted to find entities that are connected to more than two specified entities.  Here is a query-generation function for the star query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def make_star_question(types,curies,shared_type,rebuild=None):\n",
    "    \"\"\"Create a question to find entities of shared_type that are linked to all of the nodes specified in the\n",
    "    types and curies arrays.\"\"\"\n",
    "    question = {\n",
    "                #'rebuild': 'True',\n",
    "                'machine_question': {\n",
    "                    'nodes': [],\n",
    "                    'edges': []\n",
    "                }\n",
    "            }\n",
    "    if rebuild is not None and str(rebuild).upper() == 'TRUE':\n",
    "        question['rebuild']='True'\n",
    "    question['machine_question']['nodes'].append( {'id': 'n0', 'type': shared_type})\n",
    "    for i,t in enumerate(types):\n",
    "        newnode = {'id': f'n{i+1}', 'type': t}\n",
    "        if curies[i] is not None:\n",
    "            newnode['curie'] = curies[i]\n",
    "        question['machine_question']['nodes'].append(newnode)\n",
    "        question['machine_question']['edges'].append( {'id': f'e{i}', 'source_id': 'n0', 'target_id': f'n{i+1}'})\n",
    "    return question"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Suppose I have this set of GO terms, and I'd like to find genes that they all have in common:\n",
    "\n",
    "* 'voltage-gated sodium channel activity(GO:0005248)',\n",
    "* 'muscle contraction(GO:0006936)',\n",
    "* 'voltage-gated ion channel activity(GO:0005244)',\n",
    "* 'regulation of ion transmembrane transport(GO:0034765)',\n",
    "* 'sodium ion transmembrane transport(GO:0035725)',\n",
    "* 'neuronal action potential(GO:0019228)',\n",
    "* 'membrane depolarization during action potential(GO:0086010)',\n",
    "* 'sodium ion transport(GO:0006814)'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'machine_question': {'edges': [{'id': 'e0',\n",
       "    'source_id': 'n0',\n",
       "    'target_id': 'n1'},\n",
       "   {'id': 'e1', 'source_id': 'n0', 'target_id': 'n2'},\n",
       "   {'id': 'e2', 'source_id': 'n0', 'target_id': 'n3'},\n",
       "   {'id': 'e3', 'source_id': 'n0', 'target_id': 'n4'},\n",
       "   {'id': 'e4', 'source_id': 'n0', 'target_id': 'n5'},\n",
       "   {'id': 'e5', 'source_id': 'n0', 'target_id': 'n6'},\n",
       "   {'id': 'e6', 'source_id': 'n0', 'target_id': 'n7'},\n",
       "   {'id': 'e7', 'source_id': 'n0', 'target_id': 'n8'}],\n",
       "  'nodes': [{'id': 'n0', 'type': 'gene'},\n",
       "   {'curie': 'GO:0005248',\n",
       "    'id': 'n1',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0006936',\n",
       "    'id': 'n2',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0005244',\n",
       "    'id': 'n3',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0034765',\n",
       "    'id': 'n4',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0035725',\n",
       "    'id': 'n5',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0019228',\n",
       "    'id': 'n6',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0086010',\n",
       "    'id': 'n7',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0006814',\n",
       "    'id': 'n8',\n",
       "    'type': 'biological_process_or_activity'}]}}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_terms=['GO:0005248','GO:0006936','GO:0005244','GO:0034765','GO:0035725','GO:0019228','GO:0086010','GO:0006814']\n",
    "types = ['biological_process_or_activity' for g in go_terms]\n",
    "star_q = make_star_question(types,go_terms,'gene')\n",
    "star_q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "common_gene_answer = quick(star_q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n0</th>\n",
       "      <th>n1</th>\n",
       "      <th>n2</th>\n",
       "      <th>n3</th>\n",
       "      <th>n4</th>\n",
       "      <th>n5</th>\n",
       "      <th>n6</th>\n",
       "      <th>n7</th>\n",
       "      <th>n8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>GO:0005248</td>\n",
       "      <td>muscle contraction</td>\n",
       "      <td>voltage-gated ion channel activity</td>\n",
       "      <td>regulation of ion transmembrane transport</td>\n",
       "      <td>sodium ion transmembrane transport</td>\n",
       "      <td>neuronal action potential</td>\n",
       "      <td>membrane depolarization during action potential</td>\n",
       "      <td>sodium ion transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>GO:0005248</td>\n",
       "      <td>muscle contraction</td>\n",
       "      <td>voltage-gated ion channel activity</td>\n",
       "      <td>regulation of ion transmembrane transport</td>\n",
       "      <td>sodium ion transmembrane transport</td>\n",
       "      <td>neuronal action potential</td>\n",
       "      <td>membrane depolarization during action potential</td>\n",
       "      <td>sodium ion transport</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      n0          n1                  n2                                  n3  \\\n",
       "0  SCN4A  GO:0005248  muscle contraction  voltage-gated ion channel activity   \n",
       "1  SCN7A  GO:0005248  muscle contraction  voltage-gated ion channel activity   \n",
       "\n",
       "                                          n4  \\\n",
       "0  regulation of ion transmembrane transport   \n",
       "1  regulation of ion transmembrane transport   \n",
       "\n",
       "                                   n5                         n6  \\\n",
       "0  sodium ion transmembrane transport  neuronal action potential   \n",
       "1  sodium ion transmembrane transport  neuronal action potential   \n",
       "\n",
       "                                                n7                    n8  \n",
       "0  membrane depolarization during action potential  sodium ion transport  \n",
       "1  membrane depolarization during action potential  sodium ion transport  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extract_node_names(common_gene_answer)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "We can also have more than one unspecified node.  Suppose, for instance that we wanted to do the previous query, but we also wanted to know what chemicals interact with the genes that we find.  We can do another star query, where one of our spokes is unspecified:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'machine_question': {'edges': [{'id': 'e0',\n",
       "    'source_id': 'n0',\n",
       "    'target_id': 'n1'},\n",
       "   {'id': 'e1', 'source_id': 'n0', 'target_id': 'n2'},\n",
       "   {'id': 'e2', 'source_id': 'n0', 'target_id': 'n3'},\n",
       "   {'id': 'e3', 'source_id': 'n0', 'target_id': 'n4'},\n",
       "   {'id': 'e4', 'source_id': 'n0', 'target_id': 'n5'},\n",
       "   {'id': 'e5', 'source_id': 'n0', 'target_id': 'n6'},\n",
       "   {'id': 'e6', 'source_id': 'n0', 'target_id': 'n7'},\n",
       "   {'id': 'e7', 'source_id': 'n0', 'target_id': 'n8'},\n",
       "   {'id': 'e8', 'source_id': 'n0', 'target_id': 'n9'}],\n",
       "  'nodes': [{'id': 'n0', 'type': 'gene'},\n",
       "   {'curie': 'GO:0005248',\n",
       "    'id': 'n1',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0006936',\n",
       "    'id': 'n2',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0005244',\n",
       "    'id': 'n3',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0034765',\n",
       "    'id': 'n4',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0035725',\n",
       "    'id': 'n5',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0019228',\n",
       "    'id': 'n6',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0086010',\n",
       "    'id': 'n7',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0006814',\n",
       "    'id': 'n8',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'id': 'n9', 'type': 'chemical_substance'}]},\n",
       " 'rebuild': 'True'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_terms=['GO:0005248','GO:0006936','GO:0005244','GO:0034765','GO:0035725','GO:0019228','GO:0086010','GO:0006814',None]\n",
    "types = ['biological_process_or_activity' for i in range(8)]+['chemical_substance']\n",
    "star_q_compound = make_star_question(types,go_terms,'gene',rebuild=True)\n",
    "star_q_compound"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**WARNING: This takes a long time to run**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "common_gene_compound_answer = quick(star_q_compound)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n0</th>\n",
       "      <th>n9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>m-cresol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>4-chloro-3,5-dimethylphenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>4-chlorophenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>ethotoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>benzocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>tocainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>pilsicainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>lamotrigine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>ethotoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>phenazopyridine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>procainamide hydrochloride</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>phenytoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>procainamide hydrochloride</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>flecainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>levobupivacaine hydrochloride (anhydrous)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-ropivacaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>lidocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>etidocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>fospropofol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>benzocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>tocainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>zonisamide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>rufinamide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>phenazopyridine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>mepivacaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>mephenytoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>sodium(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>mexiletine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>[[4-(4-fluorophenoxy)phenyl]methylideneamino]urea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>eslicarbazepine acetate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-2-(4-chlorophenoxy)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-1-phenyl-3-(o-tolyloxy)propan-2-amine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>orphenadrine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>HWTX-IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>170 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        n0                                                 n9\n",
       "0    SCN4A                                           m-cresol\n",
       "1    SCN4A                        4-chloro-3,5-dimethylphenol\n",
       "2    SCN4A                                     4-chlorophenol\n",
       "3    SCN7A                                           ethotoin\n",
       "4    SCN7A                                         benzocaine\n",
       "5    SCN7A                                          tocainide\n",
       "6    SCN4A                                       pilsicainide\n",
       "7    SCN4A                                        lamotrigine\n",
       "8    SCN4A                                           ethotoin\n",
       "9    SCN7A                                    phenazopyridine\n",
       "10   SCN4A                         procainamide hydrochloride\n",
       "11   SCN4A                                          phenytoin\n",
       "12   SCN7A                         procainamide hydrochloride\n",
       "13   SCN4A                                         flecainide\n",
       "14   SCN4A          levobupivacaine hydrochloride (anhydrous)\n",
       "15   SCN4A                                    (S)-ropivacaine\n",
       "16   SCN4A                                          lidocaine\n",
       "17   SCN4A                                         etidocaine\n",
       "18   SCN4A                                        fospropofol\n",
       "19   SCN4A                                         benzocaine\n",
       "20   SCN4A  1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...\n",
       "21   SCN4A                                          tocainide\n",
       "22   SCN4A                                         zonisamide\n",
       "23   SCN7A                                         rufinamide\n",
       "24   SCN4A                                    phenazopyridine\n",
       "25   SCN4A                                        mepivacaine\n",
       "26   SCN7A                                        mephenytoin\n",
       "27   SCN7A                                         sodium(1+)\n",
       "28   SCN4A                                         mexiletine\n",
       "29   SCN4A  [[4-(4-fluorophenoxy)phenyl]methylideneamino]urea\n",
       "..     ...                                                ...\n",
       "140  SCN4A                            eslicarbazepine acetate\n",
       "141  SCN4A         (S)-2-(4-chlorophenoxy)-1-phenylethanamine\n",
       "142  SCN4A  (R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine\n",
       "143  SCN4A          (S)-1-phenyl-3-(o-tolyloxy)propan-2-amine\n",
       "144  SCN4A  (S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...\n",
       "145  SCN4A                                                 NA\n",
       "146  SCN4A                                                 NA\n",
       "147  SCN4A  N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...\n",
       "148  SCN4A                                                 NA\n",
       "149  SCN4A                                                 NA\n",
       "150  SCN4A                                                 NA\n",
       "151  SCN4A                                                 NA\n",
       "152  SCN4A                                                 NA\n",
       "153  SCN4A                                                 NA\n",
       "154  SCN4A                                                 NA\n",
       "155  SCN4A  3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...\n",
       "156  SCN4A                                       orphenadrine\n",
       "157  SCN4A                                                 NA\n",
       "158  SCN4A                                                 NA\n",
       "159  SCN4A     (R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine\n",
       "160  SCN4A                                                 NA\n",
       "161  SCN4A        3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole\n",
       "162  SCN4A                                                 NA\n",
       "163  SCN4A                                            HWTX-IV\n",
       "164  SCN4A                                                 NA\n",
       "165  SCN4A                                                 NA\n",
       "166  SCN4A                                                 NA\n",
       "167  SCN4A                                                 NA\n",
       "168  SCN4A                                                 NA\n",
       "169  SCN4A                                                 NA\n",
       "\n",
       "[170 rows x 2 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cgc_nodes = extract_node_names(common_gene_compound_answer)\n",
    "cgc_nodes[['n0','n9']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Many other graph shapes are possible, these are simply a couple of examples.  For instance, we could add onto this query an edge from node 0 to a new disease node (node 10) and also include an edge from the target node (node 9) to this same disease node. Now we would be finding genes that share a set of GO terms, then looking for disease,drug pairs such that the disease, the drug, and the gene are all interconnected."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Controlling the number of results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "By default, quick will return up to 250 results, but this can be raised or lowered using the max_results parameter.  We've already rebuilt this question, so we won't do rebuild=True this time, which should make it go faster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "star_q_compound = make_star_question(types,go_terms,'gene')\n",
    "common_gene_compound_answer_10 = quick(star_q_compound,max_results=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n0</th>\n",
       "      <th>n9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>m-cresol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>4-chloro-3,5-dimethylphenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>4-chlorophenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>ethotoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>benzocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>tocainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>pilsicainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>lamotrigine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>ethotoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>phenazopyridine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>procainamide hydrochloride</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>phenytoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>procainamide hydrochloride</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>flecainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>levobupivacaine hydrochloride (anhydrous)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-ropivacaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>lidocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>etidocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>fospropofol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>benzocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>tocainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>zonisamide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>rufinamide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>phenazopyridine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>mepivacaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>mephenytoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>sodium(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>mexiletine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>[[4-(4-fluorophenoxy)phenyl]methylideneamino]urea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>eslicarbazepine acetate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-2-(4-chlorophenoxy)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-1-phenyl-3-(o-tolyloxy)propan-2-amine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>orphenadrine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>HWTX-IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>170 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        n0                                                 n9\n",
       "0    SCN4A                                           m-cresol\n",
       "1    SCN4A                        4-chloro-3,5-dimethylphenol\n",
       "2    SCN4A                                     4-chlorophenol\n",
       "3    SCN7A                                           ethotoin\n",
       "4    SCN7A                                         benzocaine\n",
       "5    SCN7A                                          tocainide\n",
       "6    SCN4A                                       pilsicainide\n",
       "7    SCN4A                                        lamotrigine\n",
       "8    SCN4A                                           ethotoin\n",
       "9    SCN7A                                    phenazopyridine\n",
       "10   SCN4A                         procainamide hydrochloride\n",
       "11   SCN4A                                          phenytoin\n",
       "12   SCN7A                         procainamide hydrochloride\n",
       "13   SCN4A                                         flecainide\n",
       "14   SCN4A          levobupivacaine hydrochloride (anhydrous)\n",
       "15   SCN4A                                    (S)-ropivacaine\n",
       "16   SCN4A                                          lidocaine\n",
       "17   SCN4A                                         etidocaine\n",
       "18   SCN4A                                        fospropofol\n",
       "19   SCN4A                                         benzocaine\n",
       "20   SCN4A  1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...\n",
       "21   SCN4A                                          tocainide\n",
       "22   SCN4A                                         zonisamide\n",
       "23   SCN7A                                         rufinamide\n",
       "24   SCN4A                                    phenazopyridine\n",
       "25   SCN4A                                        mepivacaine\n",
       "26   SCN7A                                        mephenytoin\n",
       "27   SCN7A                                         sodium(1+)\n",
       "28   SCN4A                                         mexiletine\n",
       "29   SCN4A  [[4-(4-fluorophenoxy)phenyl]methylideneamino]urea\n",
       "..     ...                                                ...\n",
       "140  SCN4A                            eslicarbazepine acetate\n",
       "141  SCN4A         (S)-2-(4-chlorophenoxy)-1-phenylethanamine\n",
       "142  SCN4A  (R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine\n",
       "143  SCN4A          (S)-1-phenyl-3-(o-tolyloxy)propan-2-amine\n",
       "144  SCN4A  (S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...\n",
       "145  SCN4A                                                 NA\n",
       "146  SCN4A                                                 NA\n",
       "147  SCN4A  N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...\n",
       "148  SCN4A                                                 NA\n",
       "149  SCN4A                                                 NA\n",
       "150  SCN4A                                                 NA\n",
       "151  SCN4A                                                 NA\n",
       "152  SCN4A                                                 NA\n",
       "153  SCN4A                                                 NA\n",
       "154  SCN4A                                                 NA\n",
       "155  SCN4A  3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...\n",
       "156  SCN4A                                       orphenadrine\n",
       "157  SCN4A                                                 NA\n",
       "158  SCN4A                                                 NA\n",
       "159  SCN4A     (R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine\n",
       "160  SCN4A                                                 NA\n",
       "161  SCN4A        3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole\n",
       "162  SCN4A                                                 NA\n",
       "163  SCN4A                                            HWTX-IV\n",
       "164  SCN4A                                                 NA\n",
       "165  SCN4A                                                 NA\n",
       "166  SCN4A                                                 NA\n",
       "167  SCN4A                                                 NA\n",
       "168  SCN4A                                                 NA\n",
       "169  SCN4A                                                 NA\n",
       "\n",
       "[170 rows x 2 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cgc_nodes_10 = extract_node_names(common_gene_compound_answer_10)\n",
    "cgc_nodes_10[['n0','n9']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Rebuild and caching"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "For details on ROBOKOP caching, and guidance for when to use the `rebuild` parameter, see the notebook on the expand service.  Briefly, the quick service will only run against the cache unless the rebuild parameter is passed, in which case services are re-queried.   The rebuild parameter is passed as part of the json query as seen in the `make_star_query` function above. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Edge Properties"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "In the expand notebook, we saw how a particular predicate could be chosen to limit the results.  This same type information can in fact be added to any edge in a quick query, providing fine control to the results returned."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
