{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Using ROBOKOP's Quick Question Answering Service"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The expand, enrich, and similarity services each offer bite-sized approaches to interacting with ROBOKOP.  The ROBOKOP \"quick\" service offers a slightly more complex approach to answering questions.   We'll be using the following function to call the quick service.  Note that calling the service requires posting a question.  We'll discuss the format of the question below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "robokop_server='robokop.renci.org'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from gg_functions import parse_answer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def quick(question,max_results=None,output_format=None,max_connectivity=None):\n",
    "    url=f'http://{robokop_server}:80/api/simple/quick/'\n",
    "    if max_results is not None:\n",
    "        url += f'?max_results={max_results}'\n",
    "    if output_format is not None:\n",
    "        j = '&' if '?' in url else '?'\n",
    "        url += f'{j}output_format={output_format}'\n",
    "    if max_connectivity is not None:\n",
    "        j = '&' if '?' in url else '?'\n",
    "        url += f'{j}max_connectivity={max_connectivity}'\n",
    "    response = requests.post(url,json=question)\n",
    "    print( f\"Return Status: {response.status_code}\" )\n",
    "    if response.status_code == 200:\n",
    "        return response.json()\n",
    "    return response"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Question Format and Basic Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The question is a python dictionary.  It takes a key `machine_question`, which is a dictionary containing a list of `nodes` and a list of `edges`.   Each node object needs an integer `id`.   Any node can also have a `curie` specifying that node.  The edges define connections between the nodes using the identifiers of the nodes as source and targets. The following function shows how to construct a single-hop question.  It takes a specified node of `type1` and looks for any node of `type2`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def make_one_step_question(type1, id1, type2,rebuild = None):\n",
    "    question = {\n",
    "                'machine_question': {\n",
    "                    'nodes': [\n",
    "                        {\n",
    "                            'id': 'n0',\n",
    "                            'curie': id1,\n",
    "                            'type': type1\n",
    "                        },\n",
    "                        {\n",
    "                            'id': 'n1',\n",
    "                            'type': type2\n",
    "                        }\n",
    "                    ],\n",
    "                    'edges': [\n",
    "                        {\n",
    "                            'id': 'e0',\n",
    "                            'source_id': 'n0',\n",
    "                            'target_id': 'n1'\n",
    "                        }\n",
    "                    ]\n",
    "                }\n",
    "            }\n",
    "    if rebuild is not None and str(rebuild).upper() == 'TRUE':\n",
    "        question['rebuild'] = 'True'\n",
    "    return question"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Here, we will specify a one-hop question asking for the phenotypes associated with Fanconi Anemia (MONDO:0019391).  We first construct the question then use it to call the quick service.  In fact, this is how the expand function is implemented internally.  The output is returned in the KG standard answer format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "q = make_one_step_question('disease','MONDO:0019391','phenotypic_feature')\n",
    "r = quick(q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Uncomment to see a big long answer\n",
    "import json\n",
    "#print( json.dumps(r, indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The output can be simplified with the function parse_answer.  That function is loaded into this notebook from the gg_functions library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>e0 - edge_source</th>\n",
       "      <th>e0 - type</th>\n",
       "      <th>n1 - id</th>\n",
       "      <th>n1 - name</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001915</td>\n",
       "      <td>Aplastic anemia</td>\n",
       "      <td>1.728762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0003002</td>\n",
       "      <td>Breast carcinoma</td>\n",
       "      <td>1.725367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0004808</td>\n",
       "      <td>Acute myeloid leukemia</td>\n",
       "      <td>1.724740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0002860</td>\n",
       "      <td>Squamous cell carcinoma</td>\n",
       "      <td>1.722526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0100615</td>\n",
       "      <td>Ovarian neoplasm</td>\n",
       "      <td>1.721493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0002488</td>\n",
       "      <td>Acute leukemia</td>\n",
       "      <td>1.721270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0002721</td>\n",
       "      <td>Immunodeficiency</td>\n",
       "      <td>1.720644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000135</td>\n",
       "      <td>Hypogonadism</td>\n",
       "      <td>1.720542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0004430</td>\n",
       "      <td>Severe combined immunodeficiency</td>\n",
       "      <td>1.720526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0004810</td>\n",
       "      <td>Congenital hypoplastic anemia</td>\n",
       "      <td>1.720520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0006721</td>\n",
       "      <td>Acute lymphoblastic leukemia</td>\n",
       "      <td>1.720499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001908</td>\n",
       "      <td>Hypoplastic anemia</td>\n",
       "      <td>1.720480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001009</td>\n",
       "      <td>Telangiectasia</td>\n",
       "      <td>1.720427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000824</td>\n",
       "      <td>Growth hormone deficiency</td>\n",
       "      <td>1.720405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001251</td>\n",
       "      <td>Ataxia</td>\n",
       "      <td>1.720336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000789</td>\n",
       "      <td>Infertility</td>\n",
       "      <td>1.720310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0010972</td>\n",
       "      <td>Anemia of inadequate production</td>\n",
       "      <td>1.720194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001402</td>\n",
       "      <td>Hepatocellular carcinoma</td>\n",
       "      <td>1.720175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0100806</td>\n",
       "      <td>Sepsis</td>\n",
       "      <td>1.720099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001972</td>\n",
       "      <td>Macrocytic anemia</td>\n",
       "      <td>1.720037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000953</td>\n",
       "      <td>Hyperpigmentation of the skin</td>\n",
       "      <td>1.720036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000286</td>\n",
       "      <td>Epicanthus</td>\n",
       "      <td>1.719917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0003254</td>\n",
       "      <td>Abnormality of DNA repair</td>\n",
       "      <td>1.719881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0008064</td>\n",
       "      <td>Ichthyosis</td>\n",
       "      <td>1.719815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0005510</td>\n",
       "      <td>Transient erythroblastopenia</td>\n",
       "      <td>1.719801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0005505</td>\n",
       "      <td>Refractory anemia</td>\n",
       "      <td>1.719794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001900</td>\n",
       "      <td>Increased hemoglobin</td>\n",
       "      <td>1.719790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0011018</td>\n",
       "      <td>Abnormality of the cell cycle</td>\n",
       "      <td>1.719761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0004377</td>\n",
       "      <td>Hematological neoplasm</td>\n",
       "      <td>1.619721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0007354</td>\n",
       "      <td>Amyotrophic lateral sclerosis</td>\n",
       "      <td>1.619721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0011842</td>\n",
       "      <td>Abnormality of skeletal morphology</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0030976</td>\n",
       "      <td>Abnormal factor VIII activity</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001928</td>\n",
       "      <td>Abnormality of coagulation</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0410004</td>\n",
       "      <td>Cleft secondary palate</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0100737</td>\n",
       "      <td>Abnormal hard palate morphology</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0011875</td>\n",
       "      <td>Abnormal platelet morphology</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000202</td>\n",
       "      <td>Oral cleft</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0011338</td>\n",
       "      <td>Abnormality of mouth shape</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0010989</td>\n",
       "      <td>Abnormality of the intrinsic pathway</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000152</td>\n",
       "      <td>Abnormality of head or neck</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0003256</td>\n",
       "      <td>Abnormality of the coagulation cascade</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>biolink.phenotype_get_disease</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000153</td>\n",
       "      <td>Abnormality of the mouth</td>\n",
       "      <td>0.540111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000268</td>\n",
       "      <td>Dolichocephaly</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0006824</td>\n",
       "      <td>Cranial nerve paralysis</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0002245</td>\n",
       "      <td>Meckel diverticulum</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001347</td>\n",
       "      <td>Hyperreflexia</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000340</td>\n",
       "      <td>Sloping forehead</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001537</td>\n",
       "      <td>Umbilical hernia</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000582</td>\n",
       "      <td>Upslanted palpebral fissure</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0100587</td>\n",
       "      <td>Abnormality of the preputium</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0002414</td>\n",
       "      <td>Spina bifida</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000453</td>\n",
       "      <td>Choanal atresia</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000027</td>\n",
       "      <td>Azoospermia</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001053</td>\n",
       "      <td>Hypopigmented skin patches</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0002823</td>\n",
       "      <td>Abnormality of femur morphology</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0005344</td>\n",
       "      <td>Abnormal carotid artery morphology</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0004349</td>\n",
       "      <td>Reduced bone mineral density</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0008053</td>\n",
       "      <td>Aplasia/Hypoplasia of the iris</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0000813</td>\n",
       "      <td>Bicornuate uterus</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>biolink.disease_get_phenotype</td>\n",
       "      <td>has_phenotype</td>\n",
       "      <td>HP:0001639</td>\n",
       "      <td>Hypertrophic cardiomyopathy</td>\n",
       "      <td>0.517653</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>250 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  e0 - edge_source      e0 - type    n1 - id   \\\n",
       "0    biolink.disease_get_phenotype  has_phenotype  HP:0001915   \n",
       "1    biolink.disease_get_phenotype  has_phenotype  HP:0003002   \n",
       "2    biolink.disease_get_phenotype  has_phenotype  HP:0004808   \n",
       "3    biolink.disease_get_phenotype  has_phenotype  HP:0002860   \n",
       "4    biolink.disease_get_phenotype  has_phenotype  HP:0100615   \n",
       "5    biolink.disease_get_phenotype  has_phenotype  HP:0002488   \n",
       "6    biolink.disease_get_phenotype  has_phenotype  HP:0002721   \n",
       "7    biolink.disease_get_phenotype  has_phenotype  HP:0000135   \n",
       "8    biolink.disease_get_phenotype  has_phenotype  HP:0004430   \n",
       "9    biolink.disease_get_phenotype  has_phenotype  HP:0004810   \n",
       "10   biolink.disease_get_phenotype  has_phenotype  HP:0006721   \n",
       "11   biolink.disease_get_phenotype  has_phenotype  HP:0001908   \n",
       "12   biolink.disease_get_phenotype  has_phenotype  HP:0001009   \n",
       "13   biolink.disease_get_phenotype  has_phenotype  HP:0000824   \n",
       "14   biolink.disease_get_phenotype  has_phenotype  HP:0001251   \n",
       "15   biolink.disease_get_phenotype  has_phenotype  HP:0000789   \n",
       "16   biolink.disease_get_phenotype  has_phenotype  HP:0010972   \n",
       "17   biolink.disease_get_phenotype  has_phenotype  HP:0001402   \n",
       "18   biolink.disease_get_phenotype  has_phenotype  HP:0100806   \n",
       "19   biolink.disease_get_phenotype  has_phenotype  HP:0001972   \n",
       "20   biolink.disease_get_phenotype  has_phenotype  HP:0000953   \n",
       "21   biolink.disease_get_phenotype  has_phenotype  HP:0000286   \n",
       "22   biolink.disease_get_phenotype  has_phenotype  HP:0003254   \n",
       "23   biolink.disease_get_phenotype  has_phenotype  HP:0008064   \n",
       "24   biolink.disease_get_phenotype  has_phenotype  HP:0005510   \n",
       "25   biolink.disease_get_phenotype  has_phenotype  HP:0005505   \n",
       "26   biolink.disease_get_phenotype  has_phenotype  HP:0001900   \n",
       "27   biolink.disease_get_phenotype  has_phenotype  HP:0011018   \n",
       "28   biolink.disease_get_phenotype  has_phenotype  HP:0004377   \n",
       "29   biolink.disease_get_phenotype  has_phenotype  HP:0007354   \n",
       "..                             ...            ...         ...   \n",
       "220  biolink.phenotype_get_disease  has_phenotype  HP:0011842   \n",
       "221  biolink.phenotype_get_disease  has_phenotype  HP:0030976   \n",
       "222  biolink.phenotype_get_disease  has_phenotype  HP:0001928   \n",
       "223  biolink.phenotype_get_disease  has_phenotype  HP:0410004   \n",
       "224  biolink.phenotype_get_disease  has_phenotype  HP:0100737   \n",
       "225  biolink.phenotype_get_disease  has_phenotype  HP:0011875   \n",
       "226  biolink.phenotype_get_disease  has_phenotype  HP:0000202   \n",
       "227  biolink.phenotype_get_disease  has_phenotype  HP:0011338   \n",
       "228  biolink.phenotype_get_disease  has_phenotype  HP:0010989   \n",
       "229  biolink.phenotype_get_disease  has_phenotype  HP:0000152   \n",
       "230  biolink.phenotype_get_disease  has_phenotype  HP:0003256   \n",
       "231  biolink.phenotype_get_disease  has_phenotype  HP:0000153   \n",
       "232  biolink.disease_get_phenotype  has_phenotype  HP:0000268   \n",
       "233  biolink.disease_get_phenotype  has_phenotype  HP:0006824   \n",
       "234  biolink.disease_get_phenotype  has_phenotype  HP:0002245   \n",
       "235  biolink.disease_get_phenotype  has_phenotype  HP:0001347   \n",
       "236  biolink.disease_get_phenotype  has_phenotype  HP:0000340   \n",
       "237  biolink.disease_get_phenotype  has_phenotype  HP:0001537   \n",
       "238  biolink.disease_get_phenotype  has_phenotype  HP:0000582   \n",
       "239  biolink.disease_get_phenotype  has_phenotype  HP:0100587   \n",
       "240  biolink.disease_get_phenotype  has_phenotype  HP:0002414   \n",
       "241  biolink.disease_get_phenotype  has_phenotype  HP:0000453   \n",
       "242  biolink.disease_get_phenotype  has_phenotype  HP:0000027   \n",
       "243  biolink.disease_get_phenotype  has_phenotype  HP:0001053   \n",
       "244  biolink.disease_get_phenotype  has_phenotype  HP:0002823   \n",
       "245  biolink.disease_get_phenotype  has_phenotype  HP:0005344   \n",
       "246  biolink.disease_get_phenotype  has_phenotype  HP:0004349   \n",
       "247  biolink.disease_get_phenotype  has_phenotype  HP:0008053   \n",
       "248  biolink.disease_get_phenotype  has_phenotype  HP:0000813   \n",
       "249  biolink.disease_get_phenotype  has_phenotype  HP:0001639   \n",
       "\n",
       "                                 n1 - name      score  \n",
       "0                           Aplastic anemia  1.728762  \n",
       "1                          Breast carcinoma  1.725367  \n",
       "2                    Acute myeloid leukemia  1.724740  \n",
       "3                   Squamous cell carcinoma  1.722526  \n",
       "4                          Ovarian neoplasm  1.721493  \n",
       "5                            Acute leukemia  1.721270  \n",
       "6                          Immunodeficiency  1.720644  \n",
       "7                              Hypogonadism  1.720542  \n",
       "8          Severe combined immunodeficiency  1.720526  \n",
       "9             Congenital hypoplastic anemia  1.720520  \n",
       "10             Acute lymphoblastic leukemia  1.720499  \n",
       "11                       Hypoplastic anemia  1.720480  \n",
       "12                           Telangiectasia  1.720427  \n",
       "13                Growth hormone deficiency  1.720405  \n",
       "14                                   Ataxia  1.720336  \n",
       "15                              Infertility  1.720310  \n",
       "16          Anemia of inadequate production  1.720194  \n",
       "17                 Hepatocellular carcinoma  1.720175  \n",
       "18                                   Sepsis  1.720099  \n",
       "19                        Macrocytic anemia  1.720037  \n",
       "20            Hyperpigmentation of the skin  1.720036  \n",
       "21                               Epicanthus  1.719917  \n",
       "22                Abnormality of DNA repair  1.719881  \n",
       "23                               Ichthyosis  1.719815  \n",
       "24             Transient erythroblastopenia  1.719801  \n",
       "25                        Refractory anemia  1.719794  \n",
       "26                     Increased hemoglobin  1.719790  \n",
       "27            Abnormality of the cell cycle  1.719761  \n",
       "28                   Hematological neoplasm  1.619721  \n",
       "29            Amyotrophic lateral sclerosis  1.619721  \n",
       "..                                      ...       ...  \n",
       "220      Abnormality of skeletal morphology  0.540111  \n",
       "221           Abnormal factor VIII activity  0.540111  \n",
       "222              Abnormality of coagulation  0.540111  \n",
       "223                  Cleft secondary palate  0.540111  \n",
       "224         Abnormal hard palate morphology  0.540111  \n",
       "225            Abnormal platelet morphology  0.540111  \n",
       "226                              Oral cleft  0.540111  \n",
       "227              Abnormality of mouth shape  0.540111  \n",
       "228    Abnormality of the intrinsic pathway  0.540111  \n",
       "229             Abnormality of head or neck  0.540111  \n",
       "230  Abnormality of the coagulation cascade  0.540111  \n",
       "231                Abnormality of the mouth  0.540111  \n",
       "232                          Dolichocephaly  0.517653  \n",
       "233                 Cranial nerve paralysis  0.517653  \n",
       "234                     Meckel diverticulum  0.517653  \n",
       "235                           Hyperreflexia  0.517653  \n",
       "236                        Sloping forehead  0.517653  \n",
       "237                        Umbilical hernia  0.517653  \n",
       "238             Upslanted palpebral fissure  0.517653  \n",
       "239            Abnormality of the preputium  0.517653  \n",
       "240                            Spina bifida  0.517653  \n",
       "241                         Choanal atresia  0.517653  \n",
       "242                             Azoospermia  0.517653  \n",
       "243              Hypopigmented skin patches  0.517653  \n",
       "244         Abnormality of femur morphology  0.517653  \n",
       "245      Abnormal carotid artery morphology  0.517653  \n",
       "246            Reduced bone mineral density  0.517653  \n",
       "247          Aplasia/Hypoplasia of the iris  0.517653  \n",
       "248                       Bicornuate uterus  0.517653  \n",
       "249             Hypertrophic cardiomyopathy  0.517653  \n",
       "\n",
       "[250 rows x 5 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_answer(r)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Multi-step questions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "It's straightforward to generalize this linear query to an N-item path.  This function constructs a question from a list of types and a list of identifiers.  The types are the types of nodes that will be traversed along the path, and the identifiers represent fixed elements in the path.  The length of types and ids should be equal, and free nodes should specify an id of None:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def make_N_step_question(types,curies,rebuild = None):\n",
    "    question = {\n",
    "                'machine_question': {\n",
    "                    'nodes': [],\n",
    "                    'edges': []\n",
    "                }\n",
    "            }\n",
    "    if rebuild is not None and str(rebuild).upper() == 'TRUE':\n",
    "        question['rebuild'] = 'True'\n",
    "    ei = 0\n",
    "    for i,t in enumerate(types):\n",
    "        newnode = {'id': f'n{i}', 'type': t}\n",
    "        if curies[i] is not None:\n",
    "            newnode['curie'] = curies[i]\n",
    "        question['machine_question']['nodes'].append(newnode)\n",
    "        if i > 0:\n",
    "            question['machine_question']['edges'].append( {'id': f'e{ei}','source_id': f'n{i-1}', 'target_id': f'n{i}'})\n",
    "            ei += 1\n",
    "    return question"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "We can recapitulate our previous question with this new function like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "newq = make_N_step_question(['disease','phenotypic_feature'],['MONDO:0019391',None])\n",
    "q == newq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Now we could expand to a longer query.  The following question will start at the disease MONDO:0019391, go to a gene, and from there to a biological process or activity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"machine_question\": {\n",
      "        \"nodes\": [\n",
      "            {\n",
      "                \"id\": \"n0\",\n",
      "                \"type\": \"disease\",\n",
      "                \"curie\": \"MONDO:0019391\"\n",
      "            },\n",
      "            {\n",
      "                \"id\": \"n1\",\n",
      "                \"type\": \"gene\"\n",
      "            },\n",
      "            {\n",
      "                \"id\": \"n2\",\n",
      "                \"type\": \"biological_process_or_activity\"\n",
      "            }\n",
      "        ],\n",
      "        \"edges\": [\n",
      "            {\n",
      "                \"id\": \"e0\",\n",
      "                \"source_id\": \"n0\",\n",
      "                \"target_id\": \"n1\"\n",
      "            },\n",
      "            {\n",
      "                \"id\": \"e1\",\n",
      "                \"source_id\": \"n1\",\n",
      "                \"target_id\": \"n2\"\n",
      "            }\n",
      "        ]\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "two_step_question = make_N_step_question(['disease','gene','biological_process_or_activity'],['MONDO:0019391',None,None])\n",
    "print(json.dumps(two_step_question,indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "two_step_answer = quick(two_step_question)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "We can extract the node names along the paths with a simple function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n0 - name</th>\n",
       "      <th>n1 - name</th>\n",
       "      <th>n2 - name</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCG</td>\n",
       "      <td>response to radiation</td>\n",
       "      <td>0.558728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RAD51C</td>\n",
       "      <td>DNA recombination</td>\n",
       "      <td>0.555804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>hemopoiesis</td>\n",
       "      <td>0.544295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>tumor necrosis factor receptor binding</td>\n",
       "      <td>0.523751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCA</td>\n",
       "      <td>regulation of cell proliferation</td>\n",
       "      <td>0.493756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCL</td>\n",
       "      <td>regulation of cell proliferation</td>\n",
       "      <td>0.489189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>DNA recombination</td>\n",
       "      <td>0.489088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>MAD2L2</td>\n",
       "      <td>cell division</td>\n",
       "      <td>0.488393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>TNF</td>\n",
       "      <td>inflammatory response</td>\n",
       "      <td>0.465679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.452533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>regulation of apoptotic process</td>\n",
       "      <td>0.445331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>response to UV</td>\n",
       "      <td>0.445122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>regulation of cell proliferation</td>\n",
       "      <td>0.444966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>MX1</td>\n",
       "      <td>innate immune response</td>\n",
       "      <td>0.444392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCF</td>\n",
       "      <td>protein ubiquitination</td>\n",
       "      <td>0.444250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCC</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.441425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RAD51C</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.440971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCG</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.440748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCA</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.437623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>SLX4</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.434165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCL</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.433227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.408658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRIP1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>0.408610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCC</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.405009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCG</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.404687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>SLX4</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>0.402419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRIP1</td>\n",
       "      <td>response to toxic substance</td>\n",
       "      <td>0.401411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRIP1</td>\n",
       "      <td>double-strand break repair</td>\n",
       "      <td>0.400841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCA</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.400603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RAD51C</td>\n",
       "      <td>double-strand break repair via homologous reco...</td>\n",
       "      <td>0.399732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BLM</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>0.291906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>WRN</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>0.291764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RFWD3</td>\n",
       "      <td>chromosome breakage</td>\n",
       "      <td>0.291436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>REV1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>0.290931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>mammary gland development</td>\n",
       "      <td>0.290684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>centrosome duplication</td>\n",
       "      <td>0.290667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>single-stranded DNA binding</td>\n",
       "      <td>0.290635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>DNA2</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>0.290558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>meiotic recombination checkpoint</td>\n",
       "      <td>0.290541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>histone acetyltransferase activity</td>\n",
       "      <td>0.290538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>L3MBTL1</td>\n",
       "      <td>hemopoiesis</td>\n",
       "      <td>0.290287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>TERT</td>\n",
       "      <td>telomerase activity</td>\n",
       "      <td>0.290102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>ATRIP</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>0.289856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RMI1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>0.289697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCE</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.288822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>TP53BP1</td>\n",
       "      <td>cellular response to DNA damage stimulus</td>\n",
       "      <td>0.288814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RNF8</td>\n",
       "      <td>cellular response to DNA damage stimulus</td>\n",
       "      <td>0.287910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BLM</td>\n",
       "      <td>cellular response to DNA damage stimulus</td>\n",
       "      <td>0.287840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>WRN</td>\n",
       "      <td>cellular response to DNA damage stimulus</td>\n",
       "      <td>0.287610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>ATM</td>\n",
       "      <td>cell cycle arrest</td>\n",
       "      <td>0.287108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>DDX11</td>\n",
       "      <td>cellular response to DNA damage stimulus</td>\n",
       "      <td>0.287104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FBXO18</td>\n",
       "      <td>cellular response to DNA damage stimulus</td>\n",
       "      <td>0.287044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FAN1</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.286131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>MUS81</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.285931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>DCLRE1B</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.285811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FAAP100</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.285761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RNF8</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.285731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>DCLRE1A</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.285721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>POLN</td>\n",
       "      <td>interstrand cross-link repair</td>\n",
       "      <td>0.285691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>CENPS</td>\n",
       "      <td>cell division</td>\n",
       "      <td>0.285597</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>250 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         n0 - name  n1 - name   \\\n",
       "0    Fanconi Anemia      FANCG   \n",
       "1    Fanconi Anemia     RAD51C   \n",
       "2    Fanconi Anemia      BRCA2   \n",
       "3    Fanconi Anemia        TNF   \n",
       "4    Fanconi Anemia      FANCA   \n",
       "5    Fanconi Anemia      FANCL   \n",
       "6    Fanconi Anemia      RAD51   \n",
       "7    Fanconi Anemia     MAD2L2   \n",
       "8    Fanconi Anemia        TNF   \n",
       "9    Fanconi Anemia      RAD51   \n",
       "10   Fanconi Anemia      BRCA1   \n",
       "11   Fanconi Anemia      ERCC4   \n",
       "12   Fanconi Anemia      BRCA1   \n",
       "13   Fanconi Anemia        MX1   \n",
       "14   Fanconi Anemia      FANCF   \n",
       "15   Fanconi Anemia      FANCC   \n",
       "16   Fanconi Anemia     RAD51C   \n",
       "17   Fanconi Anemia      FANCG   \n",
       "18   Fanconi Anemia      FANCA   \n",
       "19   Fanconi Anemia       SLX4   \n",
       "20   Fanconi Anemia      FANCL   \n",
       "21   Fanconi Anemia      ERCC4   \n",
       "22   Fanconi Anemia      BRIP1   \n",
       "23   Fanconi Anemia      FANCC   \n",
       "24   Fanconi Anemia      FANCG   \n",
       "25   Fanconi Anemia       SLX4   \n",
       "26   Fanconi Anemia      BRIP1   \n",
       "27   Fanconi Anemia      BRIP1   \n",
       "28   Fanconi Anemia      FANCA   \n",
       "29   Fanconi Anemia     RAD51C   \n",
       "..              ...        ...   \n",
       "220  Fanconi Anemia        BLM   \n",
       "221  Fanconi Anemia        WRN   \n",
       "222  Fanconi Anemia      RFWD3   \n",
       "223  Fanconi Anemia       REV1   \n",
       "224  Fanconi Anemia      BRCA2   \n",
       "225  Fanconi Anemia      BRCA2   \n",
       "226  Fanconi Anemia      BRCA2   \n",
       "227  Fanconi Anemia       DNA2   \n",
       "228  Fanconi Anemia      BRCA2   \n",
       "229  Fanconi Anemia      BRCA2   \n",
       "230  Fanconi Anemia    L3MBTL1   \n",
       "231  Fanconi Anemia       TERT   \n",
       "232  Fanconi Anemia      ATRIP   \n",
       "233  Fanconi Anemia       RMI1   \n",
       "234  Fanconi Anemia      FANCE   \n",
       "235  Fanconi Anemia    TP53BP1   \n",
       "236  Fanconi Anemia       RNF8   \n",
       "237  Fanconi Anemia        BLM   \n",
       "238  Fanconi Anemia        WRN   \n",
       "239  Fanconi Anemia        ATM   \n",
       "240  Fanconi Anemia      DDX11   \n",
       "241  Fanconi Anemia     FBXO18   \n",
       "242  Fanconi Anemia       FAN1   \n",
       "243  Fanconi Anemia      MUS81   \n",
       "244  Fanconi Anemia    DCLRE1B   \n",
       "245  Fanconi Anemia    FAAP100   \n",
       "246  Fanconi Anemia       RNF8   \n",
       "247  Fanconi Anemia    DCLRE1A   \n",
       "248  Fanconi Anemia       POLN   \n",
       "249  Fanconi Anemia      CENPS   \n",
       "\n",
       "                                            n2 - name      score  \n",
       "0                                response to radiation  0.558728  \n",
       "1                                    DNA recombination  0.555804  \n",
       "2                                          hemopoiesis  0.544295  \n",
       "3               tumor necrosis factor receptor binding  0.523751  \n",
       "4                     regulation of cell proliferation  0.493756  \n",
       "5                     regulation of cell proliferation  0.489189  \n",
       "6                                    DNA recombination  0.489088  \n",
       "7                                        cell division  0.488393  \n",
       "8                                inflammatory response  0.465679  \n",
       "9                                           DNA repair  0.452533  \n",
       "10                     regulation of apoptotic process  0.445331  \n",
       "11                                      response to UV  0.445122  \n",
       "12                    regulation of cell proliferation  0.444966  \n",
       "13                              innate immune response  0.444392  \n",
       "14                              protein ubiquitination  0.444250  \n",
       "15                                          DNA repair  0.441425  \n",
       "16                                          DNA repair  0.440971  \n",
       "17                                          DNA repair  0.440748  \n",
       "18                                          DNA repair  0.437623  \n",
       "19                                          DNA repair  0.434165  \n",
       "20                                          DNA repair  0.433227  \n",
       "21                                          DNA repair  0.408658  \n",
       "22                                     DNA replication  0.408610  \n",
       "23                       interstrand cross-link repair  0.405009  \n",
       "24                       interstrand cross-link repair  0.404687  \n",
       "25                                     DNA replication  0.402419  \n",
       "26                         response to toxic substance  0.401411  \n",
       "27                          double-strand break repair  0.400841  \n",
       "28                       interstrand cross-link repair  0.400603  \n",
       "29   double-strand break repair via homologous reco...  0.399732  \n",
       "..                                                 ...       ...  \n",
       "220                                    DNA replication  0.291906  \n",
       "221                                    DNA replication  0.291764  \n",
       "222                                chromosome breakage  0.291436  \n",
       "223                                    DNA replication  0.290931  \n",
       "224                          mammary gland development  0.290684  \n",
       "225                             centrosome duplication  0.290667  \n",
       "226                        single-stranded DNA binding  0.290635  \n",
       "227                                    DNA replication  0.290558  \n",
       "228                   meiotic recombination checkpoint  0.290541  \n",
       "229                 histone acetyltransferase activity  0.290538  \n",
       "230                                        hemopoiesis  0.290287  \n",
       "231                                telomerase activity  0.290102  \n",
       "232                                    DNA replication  0.289856  \n",
       "233                                    DNA replication  0.289697  \n",
       "234                      interstrand cross-link repair  0.288822  \n",
       "235           cellular response to DNA damage stimulus  0.288814  \n",
       "236           cellular response to DNA damage stimulus  0.287910  \n",
       "237           cellular response to DNA damage stimulus  0.287840  \n",
       "238           cellular response to DNA damage stimulus  0.287610  \n",
       "239                                  cell cycle arrest  0.287108  \n",
       "240           cellular response to DNA damage stimulus  0.287104  \n",
       "241           cellular response to DNA damage stimulus  0.287044  \n",
       "242                      interstrand cross-link repair  0.286131  \n",
       "243                      interstrand cross-link repair  0.285931  \n",
       "244                      interstrand cross-link repair  0.285811  \n",
       "245                      interstrand cross-link repair  0.285761  \n",
       "246                      interstrand cross-link repair  0.285731  \n",
       "247                      interstrand cross-link repair  0.285721  \n",
       "248                      interstrand cross-link repair  0.285691  \n",
       "249                                      cell division  0.285597  \n",
       "\n",
       "[250 rows x 4 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "two_step_nodes = parse_answer(two_step_answer,node_list=['n0','n1','n2'],edge_list=[],node_properties=['name'])\n",
    "two_step_nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Visualizing the output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "In addition to parsing the output, we can actually visualize it in robokop.  First, we need to send the quick output to the viewer, and we will return an identifier for the visualization:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "view_post_url = f'http://{robokop_server}/api/simple/view/'\n",
    "view_post_response = requests.post(view_post_url, json=two_step_answer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<a href=http://robokop.renci.org/simple/view/98e2089c-1d47-40d6-8fce-13e5663b0a65>View Answer</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "uid=view_post_response.json()\n",
    "view_url = f'http://{robokop_server}/simple/view/{uid}'\n",
    "display(HTML(f'<a href={view_url}>View Answer</a>'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Output Format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Version 0.9, output_format=message"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The results above are returned in v0.9 of the Translator standard for graph answers.  In this format, one graph is returned, containing all of the nodes and edges required for any of the answers.  Then the list of answers is given, binding nodes and edges in the question definition to nodes and edges in the shared knowledge graph.  This allows greater efficiency and specificity in relating outputs to question parameters.\n",
    "\n",
    "The response has 3 keys returning\n",
    "1. The json formatted question that was asked \n",
    "2. The shared knowledge graph \n",
    "3. The answers as defined on that graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['question_graph', 'knowledge_graph', 'answers'])\n"
     ]
    }
   ],
   "source": [
    "print(two_step_answer.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The knowledge graph consists of nodes (containing all the information about the entity) and edges (containing information about the relationship between entities)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['nodes', 'edges'])\n"
     ]
    }
   ],
   "source": [
    "print(two_step_answer['knowledge_graph'].keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Here's what the first 2 nodes and the first edge look like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "    {\n",
      "        \"congenital abnormality\": true,\n",
      "        \"name\": \"Fanconi anemia\",\n",
      "        \"nutritional or metabolic disease\": true,\n",
      "        \"id\": \"MONDO:0019391\",\n",
      "        \"equivalent_identifiers\": [\n",
      "            \"MESH:D005199\",\n",
      "            \"MEDDRA:10016218\",\n",
      "            \"MONDO:0019391\",\n",
      "            \"MEDDRA:10055206\",\n",
      "            \"UMLS:C0015625\",\n",
      "            \"ORPHANET:84\",\n",
      "            \"DOID:13636\"\n",
      "        ],\n",
      "        \"type\": \"disease\",\n",
      "        \"rare disease\": true,\n",
      "        \"syndromic disease\": true\n",
      "    },\n",
      "    {\n",
      "        \"name\": \"UBE2T\",\n",
      "        \"location\": \"1q32.1\",\n",
      "        \"locus_group\": \"protein-coding gene\",\n",
      "        \"gene_family\": [\n",
      "            \"Ubiquitin conjugating enzymes E2\",\n",
      "            \"FA complementation groups\"\n",
      "        ],\n",
      "        \"chromosome\": \"1\",\n",
      "        \"id\": \"HGNC:25009\",\n",
      "        \"equivalent_identifiers\": [\n",
      "            \"UniProtKB:Q9NPD8\",\n",
      "            \"NCBIGene:29089\",\n",
      "            \"HGNC:25009\",\n",
      "            \"ENSEMBL:ENSG00000077152\",\n",
      "            \"HGNC.SYMBOL:UBE2T\",\n",
      "            \"UniProtKB:Q9NPD8\"\n",
      "        ],\n",
      "        \"taxon\": \"9606\",\n",
      "        \"gene_family_id\": [\n",
      "            102,\n",
      "            548\n",
      "        ],\n",
      "        \"type\": \"gene\"\n",
      "    }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(two_step_answer['knowledge_graph']['nodes'][:2],indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"relation\": \"PHAROS:gene_involved\",\n",
      "    \"target_id\": \"HGNC:25009\",\n",
      "    \"edge_source\": \"pharos.disease_get_gene\",\n",
      "    \"publications\": [],\n",
      "    \"id\": \"4257632\",\n",
      "    \"predicate_id\": \"NCIT:R176\",\n",
      "    \"source_database\": \"pharos\",\n",
      "    \"source_id\": \"MONDO:0019391\",\n",
      "    \"type\": \"disease_to_gene_association\",\n",
      "    \"ctime\": 1543738834.740142,\n",
      "    \"relation_label\": \"gene_involved\",\n",
      "    \"weight\": 0.25832639515782074\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(two_step_answer['knowledge_graph']['edges'][1],indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Here's what the first answer looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"node_bindings\": {\n",
      "        \"n0\": \"MONDO:0019391\",\n",
      "        \"n1\": \"HGNC:20473\",\n",
      "        \"n2\": \"GO:0009636\"\n",
      "    },\n",
      "    \"edge_bindings\": {\n",
      "        \"e0\": [\n",
      "            \"4257531\",\n",
      "            \"6938566\",\n",
      "            \"6938600\"\n",
      "        ],\n",
      "        \"e1\": [\n",
      "            \"4294578\",\n",
      "            \"1094816\"\n",
      "        ]\n",
      "    },\n",
      "    \"score\": 0.3959587237470695\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(two_step_answer['answers'][0],indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The answers are ranked by score, so this is the highest ranking answer, and the score is given.  The `node_bindings` relate the identifiers of the questions (`n0`, `n1`, `n2`) to the identifiers in the knowledge graph (`MONDO:0019391`, `HGNC:20473`, `GO:0009636`) and we could use these identifiers to look up further information about those nodes in the graph.\n",
    "\n",
    "Similarly, the `e0` edge (connecting `n0` and `n1` in the question), is mapped to several possible edges in the graph, and we could look up the edge information using the identifier.  For instance, we can find the first edge between for `e0`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.3959587237470695,\n",
       " 0.39562714985581743,\n",
       " 0.3930911628130927,\n",
       " 0.38680121546632074,\n",
       " 0.38680121546632074,\n",
       " 0.38237448152146924,\n",
       " 0.37742069307012543,\n",
       " 0.37742069307012543,\n",
       " 0.37742069307012543,\n",
       " 0.35967233858255104,\n",
       " 0.35967233858255104,\n",
       " 0.3328828490620355,\n",
       " 0.3328828490620355,\n",
       " 0.3328828490620355,\n",
       " 0.3328828490620355,\n",
       " 0.3328828490620355,\n",
       " 0.330896351413593,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.29235569718019727,\n",
       " 0.22921193146495913,\n",
       " 0.22921193146495913,\n",
       " 0.22921193146495913,\n",
       " 0.22921193146495913,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.225403541755474,\n",
       " 0.2252964653481522,\n",
       " 0.2252964653481522,\n",
       " 0.2252964653481522,\n",
       " 0.2252964653481522,\n",
       " 0.2252964653481522,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22447493376868408,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22241744714294412,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.22095220849523287,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.2192953510262118,\n",
       " 0.21598706393112438,\n",
       " 0.21598706393112438,\n",
       " 0.21598706393112438,\n",
       " 0.21598706393112438,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.2132057566198872,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466,\n",
       " 0.20353333074854466]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scores = [a['score'] for a in two_step_answer['answers']]\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"relation\": \"PHAROS:gene_involved\",\n",
      "    \"target_id\": \"HGNC:20473\",\n",
      "    \"edge_source\": \"pharos.disease_get_gene\",\n",
      "    \"publications\": [],\n",
      "    \"id\": \"4257531\",\n",
      "    \"predicate_id\": \"NCIT:R176\",\n",
      "    \"source_database\": \"pharos\",\n",
      "    \"source_id\": \"MONDO:0019391\",\n",
      "    \"type\": \"disease_to_gene_association\",\n",
      "    \"ctime\": 1543738834.73799,\n",
      "    \"relation_label\": \"gene_involved\",\n",
      "    \"weight\": 0.25832639515782074\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "my_edge = list(filter(lambda x: x['id'] == \"4257531\", two_step_answer['knowledge_graph']['edges']))[0]\n",
    "print(json.dumps(my_edge,indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### v0.8, output_format=dense"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "For backwards compatibility, output can also be returned in v0.8 of the KG format.  To return output in this format, pass the argument `output_format=dense` to the url.   The `output_format` defaults to `message` which specifies v0.9."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "two_step_answer_dense = quick(two_step_question,output_format='dense')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['datetime', 'id', 'message', 'response_code', 'result_list'])"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "two_step_answer_dense.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'confidence': 0.3959587237470695,\n",
       " 'id': '5939821d-f96e-442c-ab37-059ffe1087e4',\n",
       " 'result_graph': {'edge_list': [{'confidence': 0.25832639515782074,\n",
       "    'num_publications': None,\n",
       "    'provided_by': 'pharos.disease_get_gene',\n",
       "    'publications': '',\n",
       "    'source_id': 'MONDO:0019391',\n",
       "    'target_id': 'HGNC:20473',\n",
       "    'type': 'disease_to_gene_association'},\n",
       "   {'confidence': 0.4071474314830641,\n",
       "    'num_publications': None,\n",
       "    'provided_by': 'biolink.disease_get_gene',\n",
       "    'publications': 'PMID:20639400',\n",
       "    'source_id': 'MONDO:0019391',\n",
       "    'target_id': 'HGNC:20473',\n",
       "    'type': 'biomarker_for'},\n",
       "   {'confidence': 0.9998780282217139,\n",
       "    'num_publications': None,\n",
       "    'provided_by': 'biolink.disease_get_gene',\n",
       "    'publications': 'PMID:25980754,PMID:19127258,PMID:16116423,PMID:16153896,PMID:16116424,PMID:27498913,PMID:22006311,PMID:24556621,PMID:26689913,PMID:26681312,PMID:17033622,PMID:26720728,PMID:25186627,PMID:25452441,PMID:26921362,PMID:21964575,PMID:26786923,PMID:21127055,PMID:24763289,PMID:26845104,PMID:27153395,PMID:18628483,PMID:20639400,PMID:26315354,PMID:22792074,PMID:20159562',\n",
       "    'source_id': 'MONDO:0019391',\n",
       "    'target_id': 'HGNC:20473',\n",
       "    'type': 'gene_associated_with_condition'},\n",
       "   {'confidence': 0.25832639515782074,\n",
       "    'num_publications': None,\n",
       "    'provided_by': 'biolink.gene_get_process_or_function',\n",
       "    'publications': '',\n",
       "    'source_id': 'HGNC:20473',\n",
       "    'target_id': 'GO:0009636',\n",
       "    'type': 'actively_involved_in'},\n",
       "   {'confidence': 0.25832639515782074,\n",
       "    'num_publications': None,\n",
       "    'provided_by': 'quickgo.go_term_to_gene_annotation',\n",
       "    'publications': '',\n",
       "    'source_id': 'HGNC:20473',\n",
       "    'target_id': 'GO:0009636',\n",
       "    'type': 'actively_involved_in'}],\n",
       "  'node_list': [{'description': 'Fanconi anemia',\n",
       "    'id': 'MONDO:0019391',\n",
       "    'name': 'Fanconi anemia',\n",
       "    'type': 'disease'},\n",
       "   {'description': 'BRIP1',\n",
       "    'id': 'HGNC:20473',\n",
       "    'name': 'BRIP1',\n",
       "    'type': 'gene'},\n",
       "   {'description': 'response to toxic substance',\n",
       "    'id': 'GO:0009636',\n",
       "    'name': 'response to toxic substance',\n",
       "    'type': 'biological_process_or_activity'}]},\n",
       " 'result_type': 'individual query answer',\n",
       " 'text': 'response to toxic substance'}"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "two_step_answer_dense['result_list'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Specifying multiple fixed nodes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The examples above have all started from one known node and expanded from it, sometimes in multiple steps.  It's also possible to have more than one node specified.  For instance, if we wanted to look for genes that link Fanconi Aneima (MONDO:0019391) and DNA repair (GO:0006281), we could rerun our previous query, but setting the final curie, like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "two_step_question_fixed_ends = \\\n",
    "   make_N_step_question(['disease','gene','biological_process_or_activity'],['MONDO:0019391',None,'GO:0006281'])\n",
    "two_step_answer_fixed_ends = quick(two_step_question_fixed_ends)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n0 - name</th>\n",
       "      <th>n1 - name</th>\n",
       "      <th>n2 - name</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.452533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCC</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.441425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RAD51C</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.440971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCG</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.440748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCA</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.437623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>SLX4</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.434165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCL</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.433227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>ERCC4</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.408658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>XRCC2</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.399073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FANCD2</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.393783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>UBE2T</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.385278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>XPA</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.327133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BLM</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.323795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>WRN</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.322999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>MUS81</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.321773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RAD51D</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.320855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>FAN1</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.320700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>RAD51AP1</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.320261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>DDX11</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.320211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>CENPS</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.270052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>SLX1B</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.270052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>SLX1B</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.270052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Fanconi Anemia</td>\n",
       "      <td>BOD1L1</td>\n",
       "      <td>DNA repair</td>\n",
       "      <td>0.270052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        n0 - name  n1 - name   n2 - name      score\n",
       "0   Fanconi Anemia      RAD51  DNA repair  0.452533\n",
       "1   Fanconi Anemia      FANCC  DNA repair  0.441425\n",
       "2   Fanconi Anemia     RAD51C  DNA repair  0.440971\n",
       "3   Fanconi Anemia      FANCG  DNA repair  0.440748\n",
       "4   Fanconi Anemia      FANCA  DNA repair  0.437623\n",
       "5   Fanconi Anemia       SLX4  DNA repair  0.434165\n",
       "6   Fanconi Anemia      FANCL  DNA repair  0.433227\n",
       "7   Fanconi Anemia      ERCC4  DNA repair  0.408658\n",
       "8   Fanconi Anemia      XRCC2  DNA repair  0.399073\n",
       "9   Fanconi Anemia     FANCD2  DNA repair  0.393783\n",
       "10  Fanconi Anemia      UBE2T  DNA repair  0.385278\n",
       "11  Fanconi Anemia        XPA  DNA repair  0.327133\n",
       "12  Fanconi Anemia        BLM  DNA repair  0.323795\n",
       "13  Fanconi Anemia        WRN  DNA repair  0.322999\n",
       "14  Fanconi Anemia      MUS81  DNA repair  0.321773\n",
       "15  Fanconi Anemia     RAD51D  DNA repair  0.320855\n",
       "16  Fanconi Anemia       FAN1  DNA repair  0.320700\n",
       "17  Fanconi Anemia   RAD51AP1  DNA repair  0.320261\n",
       "18  Fanconi Anemia      DDX11  DNA repair  0.320211\n",
       "19  Fanconi Anemia      CENPS  DNA repair  0.270052\n",
       "20  Fanconi Anemia      SLX1B  DNA repair  0.270052\n",
       "21  Fanconi Anemia      SLX1B  DNA repair  0.270052\n",
       "22  Fanconi Anemia     BOD1L1  DNA repair  0.270052"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_answer(two_step_answer_fixed_ends,node_list=['n0','n1','n2'],edge_list=[],node_properties=['name'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Non-path queries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "So far, we've only looked at linear paths.  But the question format is actually more general than that - we can define a path pattern generally. So for instance, in the above query, we find all genes that are linked to both FA and DNA repair.  But what if we wanted to find entities that are connected to more than two specified entities.  Here is a query-generation function for the star query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def make_star_question(types,curies,shared_type,rebuild=None):\n",
    "    \"\"\"Create a question to find entities of shared_type that are linked to all of the nodes specified in the\n",
    "    types and curies arrays.\"\"\"\n",
    "    question = {\n",
    "                #'rebuild': 'True',\n",
    "                'machine_question': {\n",
    "                    'nodes': [],\n",
    "                    'edges': []\n",
    "                }\n",
    "            }\n",
    "    if rebuild is not None and str(rebuild).upper() == 'TRUE':\n",
    "        question['rebuild']='True'\n",
    "    question['machine_question']['nodes'].append( {'id': 'n0', 'type': shared_type})\n",
    "    for i,t in enumerate(types):\n",
    "        newnode = {'id': f'n{i+1}', 'type': t}\n",
    "        if curies[i] is not None:\n",
    "            newnode['curie'] = curies[i]\n",
    "        question['machine_question']['nodes'].append(newnode)\n",
    "        question['machine_question']['edges'].append( {'id': f'e{i}', 'source_id': 'n0', 'target_id': f'n{i+1}'})\n",
    "    return question"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Suppose I have this set of GO terms, and I'd like to find genes that they all have in common:\n",
    "\n",
    "* 'voltage-gated sodium channel activity(GO:0005248)',\n",
    "* 'muscle contraction(GO:0006936)',\n",
    "* 'voltage-gated ion channel activity(GO:0005244)',\n",
    "* 'regulation of ion transmembrane transport(GO:0034765)',\n",
    "* 'sodium ion transmembrane transport(GO:0035725)',\n",
    "* 'neuronal action potential(GO:0019228)',\n",
    "* 'membrane depolarization during action potential(GO:0086010)',\n",
    "* 'sodium ion transport(GO:0006814)'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'machine_question': {'edges': [{'id': 'e0',\n",
       "    'source_id': 'n0',\n",
       "    'target_id': 'n1'},\n",
       "   {'id': 'e1', 'source_id': 'n0', 'target_id': 'n2'},\n",
       "   {'id': 'e2', 'source_id': 'n0', 'target_id': 'n3'},\n",
       "   {'id': 'e3', 'source_id': 'n0', 'target_id': 'n4'},\n",
       "   {'id': 'e4', 'source_id': 'n0', 'target_id': 'n5'},\n",
       "   {'id': 'e5', 'source_id': 'n0', 'target_id': 'n6'},\n",
       "   {'id': 'e6', 'source_id': 'n0', 'target_id': 'n7'},\n",
       "   {'id': 'e7', 'source_id': 'n0', 'target_id': 'n8'}],\n",
       "  'nodes': [{'id': 'n0', 'type': 'gene'},\n",
       "   {'curie': 'GO:0005248',\n",
       "    'id': 'n1',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0006936',\n",
       "    'id': 'n2',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0005244',\n",
       "    'id': 'n3',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0034765',\n",
       "    'id': 'n4',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0035725',\n",
       "    'id': 'n5',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0019228',\n",
       "    'id': 'n6',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0086010',\n",
       "    'id': 'n7',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0006814',\n",
       "    'id': 'n8',\n",
       "    'type': 'biological_process_or_activity'}]}}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_terms=['GO:0005248','GO:0006936','GO:0005244','GO:0034765','GO:0035725','GO:0019228','GO:0086010','GO:0006814']\n",
    "types = ['biological_process_or_activity' for g in go_terms]\n",
    "star_q = make_star_question(types,go_terms,'gene')\n",
    "star_q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "common_gene_answer = quick(star_q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n0</th>\n",
       "      <th>n1</th>\n",
       "      <th>n2</th>\n",
       "      <th>n3</th>\n",
       "      <th>n4</th>\n",
       "      <th>n5</th>\n",
       "      <th>n6</th>\n",
       "      <th>n7</th>\n",
       "      <th>n8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>GO:0005248</td>\n",
       "      <td>muscle contraction</td>\n",
       "      <td>voltage-gated ion channel activity</td>\n",
       "      <td>regulation of ion transmembrane transport</td>\n",
       "      <td>sodium ion transmembrane transport</td>\n",
       "      <td>neuronal action potential</td>\n",
       "      <td>membrane depolarization during action potential</td>\n",
       "      <td>sodium ion transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>GO:0005248</td>\n",
       "      <td>muscle contraction</td>\n",
       "      <td>voltage-gated ion channel activity</td>\n",
       "      <td>regulation of ion transmembrane transport</td>\n",
       "      <td>sodium ion transmembrane transport</td>\n",
       "      <td>neuronal action potential</td>\n",
       "      <td>membrane depolarization during action potential</td>\n",
       "      <td>sodium ion transport</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      n0          n1                  n2                                  n3  \\\n",
       "0  SCN4A  GO:0005248  muscle contraction  voltage-gated ion channel activity   \n",
       "1  SCN7A  GO:0005248  muscle contraction  voltage-gated ion channel activity   \n",
       "\n",
       "                                          n4  \\\n",
       "0  regulation of ion transmembrane transport   \n",
       "1  regulation of ion transmembrane transport   \n",
       "\n",
       "                                   n5                         n6  \\\n",
       "0  sodium ion transmembrane transport  neuronal action potential   \n",
       "1  sodium ion transmembrane transport  neuronal action potential   \n",
       "\n",
       "                                                n7                    n8  \n",
       "0  membrane depolarization during action potential  sodium ion transport  \n",
       "1  membrane depolarization during action potential  sodium ion transport  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extract_node_names(common_gene_answer)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "We can also have more than one unspecified node.  Suppose, for instance that we wanted to do the previous query, but we also wanted to know what chemicals interact with the genes that we find.  We can do another star query, where one of our spokes is unspecified:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'machine_question': {'edges': [{'id': 'e0',\n",
       "    'source_id': 'n0',\n",
       "    'target_id': 'n1'},\n",
       "   {'id': 'e1', 'source_id': 'n0', 'target_id': 'n2'},\n",
       "   {'id': 'e2', 'source_id': 'n0', 'target_id': 'n3'},\n",
       "   {'id': 'e3', 'source_id': 'n0', 'target_id': 'n4'},\n",
       "   {'id': 'e4', 'source_id': 'n0', 'target_id': 'n5'},\n",
       "   {'id': 'e5', 'source_id': 'n0', 'target_id': 'n6'},\n",
       "   {'id': 'e6', 'source_id': 'n0', 'target_id': 'n7'},\n",
       "   {'id': 'e7', 'source_id': 'n0', 'target_id': 'n8'},\n",
       "   {'id': 'e8', 'source_id': 'n0', 'target_id': 'n9'}],\n",
       "  'nodes': [{'id': 'n0', 'type': 'gene'},\n",
       "   {'curie': 'GO:0005248',\n",
       "    'id': 'n1',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0006936',\n",
       "    'id': 'n2',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0005244',\n",
       "    'id': 'n3',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0034765',\n",
       "    'id': 'n4',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0035725',\n",
       "    'id': 'n5',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0019228',\n",
       "    'id': 'n6',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0086010',\n",
       "    'id': 'n7',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'curie': 'GO:0006814',\n",
       "    'id': 'n8',\n",
       "    'type': 'biological_process_or_activity'},\n",
       "   {'id': 'n9', 'type': 'chemical_substance'}]},\n",
       " 'rebuild': 'True'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_terms=['GO:0005248','GO:0006936','GO:0005244','GO:0034765','GO:0035725','GO:0019228','GO:0086010','GO:0006814',None]\n",
    "types = ['biological_process_or_activity' for i in range(8)]+['chemical_substance']\n",
    "star_q_compound = make_star_question(types,go_terms,'gene',rebuild=True)\n",
    "star_q_compound"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**WARNING: This takes a long time to run**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "common_gene_compound_answer = quick(star_q_compound)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n0</th>\n",
       "      <th>n9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>m-cresol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>4-chloro-3,5-dimethylphenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>4-chlorophenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>ethotoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>benzocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>tocainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>pilsicainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>lamotrigine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>ethotoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>phenazopyridine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>procainamide hydrochloride</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>phenytoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>procainamide hydrochloride</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>flecainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>levobupivacaine hydrochloride (anhydrous)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-ropivacaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>lidocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>etidocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>fospropofol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>benzocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>tocainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>zonisamide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>rufinamide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>phenazopyridine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>mepivacaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>mephenytoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>sodium(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>mexiletine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>[[4-(4-fluorophenoxy)phenyl]methylideneamino]urea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>eslicarbazepine acetate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-2-(4-chlorophenoxy)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-1-phenyl-3-(o-tolyloxy)propan-2-amine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>orphenadrine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>HWTX-IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>170 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        n0                                                 n9\n",
       "0    SCN4A                                           m-cresol\n",
       "1    SCN4A                        4-chloro-3,5-dimethylphenol\n",
       "2    SCN4A                                     4-chlorophenol\n",
       "3    SCN7A                                           ethotoin\n",
       "4    SCN7A                                         benzocaine\n",
       "5    SCN7A                                          tocainide\n",
       "6    SCN4A                                       pilsicainide\n",
       "7    SCN4A                                        lamotrigine\n",
       "8    SCN4A                                           ethotoin\n",
       "9    SCN7A                                    phenazopyridine\n",
       "10   SCN4A                         procainamide hydrochloride\n",
       "11   SCN4A                                          phenytoin\n",
       "12   SCN7A                         procainamide hydrochloride\n",
       "13   SCN4A                                         flecainide\n",
       "14   SCN4A          levobupivacaine hydrochloride (anhydrous)\n",
       "15   SCN4A                                    (S)-ropivacaine\n",
       "16   SCN4A                                          lidocaine\n",
       "17   SCN4A                                         etidocaine\n",
       "18   SCN4A                                        fospropofol\n",
       "19   SCN4A                                         benzocaine\n",
       "20   SCN4A  1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...\n",
       "21   SCN4A                                          tocainide\n",
       "22   SCN4A                                         zonisamide\n",
       "23   SCN7A                                         rufinamide\n",
       "24   SCN4A                                    phenazopyridine\n",
       "25   SCN4A                                        mepivacaine\n",
       "26   SCN7A                                        mephenytoin\n",
       "27   SCN7A                                         sodium(1+)\n",
       "28   SCN4A                                         mexiletine\n",
       "29   SCN4A  [[4-(4-fluorophenoxy)phenyl]methylideneamino]urea\n",
       "..     ...                                                ...\n",
       "140  SCN4A                            eslicarbazepine acetate\n",
       "141  SCN4A         (S)-2-(4-chlorophenoxy)-1-phenylethanamine\n",
       "142  SCN4A  (R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine\n",
       "143  SCN4A          (S)-1-phenyl-3-(o-tolyloxy)propan-2-amine\n",
       "144  SCN4A  (S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...\n",
       "145  SCN4A                                                 NA\n",
       "146  SCN4A                                                 NA\n",
       "147  SCN4A  N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...\n",
       "148  SCN4A                                                 NA\n",
       "149  SCN4A                                                 NA\n",
       "150  SCN4A                                                 NA\n",
       "151  SCN4A                                                 NA\n",
       "152  SCN4A                                                 NA\n",
       "153  SCN4A                                                 NA\n",
       "154  SCN4A                                                 NA\n",
       "155  SCN4A  3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...\n",
       "156  SCN4A                                       orphenadrine\n",
       "157  SCN4A                                                 NA\n",
       "158  SCN4A                                                 NA\n",
       "159  SCN4A     (R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine\n",
       "160  SCN4A                                                 NA\n",
       "161  SCN4A        3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole\n",
       "162  SCN4A                                                 NA\n",
       "163  SCN4A                                            HWTX-IV\n",
       "164  SCN4A                                                 NA\n",
       "165  SCN4A                                                 NA\n",
       "166  SCN4A                                                 NA\n",
       "167  SCN4A                                                 NA\n",
       "168  SCN4A                                                 NA\n",
       "169  SCN4A                                                 NA\n",
       "\n",
       "[170 rows x 2 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cgc_nodes = extract_node_names(common_gene_compound_answer)\n",
    "cgc_nodes[['n0','n9']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Many other graph shapes are possible, these are simply a couple of examples.  For instance, we could add onto this query an edge from node 0 to a new disease node (node 10) and also include an edge from the target node (node 9) to this same disease node. Now we would be finding genes that share a set of GO terms, then looking for disease,drug pairs such that the disease, the drug, and the gene are all interconnected."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Controlling the number of results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "By default, quick will return up to 250 results, but this can be raised or lowered using the max_results parameter.  We've already rebuilt this question, so we won't do rebuild=True this time, which should make it go faster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "star_q_compound = make_star_question(types,go_terms,'gene')\n",
    "common_gene_compound_answer_10 = quick(star_q_compound,max_results=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n0</th>\n",
       "      <th>n9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>m-cresol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>4-chloro-3,5-dimethylphenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>4-chlorophenol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>ethotoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>benzocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>tocainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>pilsicainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>lamotrigine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>ethotoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>phenazopyridine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>procainamide hydrochloride</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>phenytoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>procainamide hydrochloride</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>flecainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>levobupivacaine hydrochloride (anhydrous)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-ropivacaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>lidocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>etidocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>fospropofol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>benzocaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>tocainide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>zonisamide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>rufinamide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>phenazopyridine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>mepivacaine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>mephenytoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SCN7A</td>\n",
       "      <td>sodium(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>mexiletine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>[[4-(4-fluorophenoxy)phenyl]methylideneamino]urea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>eslicarbazepine acetate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-2-(4-chlorophenoxy)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-1-phenyl-3-(o-tolyloxy)propan-2-amine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>orphenadrine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>(R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>HWTX-IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>SCN4A</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>170 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        n0                                                 n9\n",
       "0    SCN4A                                           m-cresol\n",
       "1    SCN4A                        4-chloro-3,5-dimethylphenol\n",
       "2    SCN4A                                     4-chlorophenol\n",
       "3    SCN7A                                           ethotoin\n",
       "4    SCN7A                                         benzocaine\n",
       "5    SCN7A                                          tocainide\n",
       "6    SCN4A                                       pilsicainide\n",
       "7    SCN4A                                        lamotrigine\n",
       "8    SCN4A                                           ethotoin\n",
       "9    SCN7A                                    phenazopyridine\n",
       "10   SCN4A                         procainamide hydrochloride\n",
       "11   SCN4A                                          phenytoin\n",
       "12   SCN7A                         procainamide hydrochloride\n",
       "13   SCN4A                                         flecainide\n",
       "14   SCN4A          levobupivacaine hydrochloride (anhydrous)\n",
       "15   SCN4A                                    (S)-ropivacaine\n",
       "16   SCN4A                                          lidocaine\n",
       "17   SCN4A                                         etidocaine\n",
       "18   SCN4A                                        fospropofol\n",
       "19   SCN4A                                         benzocaine\n",
       "20   SCN4A  1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...\n",
       "21   SCN4A                                          tocainide\n",
       "22   SCN4A                                         zonisamide\n",
       "23   SCN7A                                         rufinamide\n",
       "24   SCN4A                                    phenazopyridine\n",
       "25   SCN4A                                        mepivacaine\n",
       "26   SCN7A                                        mephenytoin\n",
       "27   SCN7A                                         sodium(1+)\n",
       "28   SCN4A                                         mexiletine\n",
       "29   SCN4A  [[4-(4-fluorophenoxy)phenyl]methylideneamino]urea\n",
       "..     ...                                                ...\n",
       "140  SCN4A                            eslicarbazepine acetate\n",
       "141  SCN4A         (S)-2-(4-chlorophenoxy)-1-phenylethanamine\n",
       "142  SCN4A  (R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine\n",
       "143  SCN4A          (S)-1-phenyl-3-(o-tolyloxy)propan-2-amine\n",
       "144  SCN4A  (S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...\n",
       "145  SCN4A                                                 NA\n",
       "146  SCN4A                                                 NA\n",
       "147  SCN4A  N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...\n",
       "148  SCN4A                                                 NA\n",
       "149  SCN4A                                                 NA\n",
       "150  SCN4A                                                 NA\n",
       "151  SCN4A                                                 NA\n",
       "152  SCN4A                                                 NA\n",
       "153  SCN4A                                                 NA\n",
       "154  SCN4A                                                 NA\n",
       "155  SCN4A  3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...\n",
       "156  SCN4A                                       orphenadrine\n",
       "157  SCN4A                                                 NA\n",
       "158  SCN4A                                                 NA\n",
       "159  SCN4A     (R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine\n",
       "160  SCN4A                                                 NA\n",
       "161  SCN4A        3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole\n",
       "162  SCN4A                                                 NA\n",
       "163  SCN4A                                            HWTX-IV\n",
       "164  SCN4A                                                 NA\n",
       "165  SCN4A                                                 NA\n",
       "166  SCN4A                                                 NA\n",
       "167  SCN4A                                                 NA\n",
       "168  SCN4A                                                 NA\n",
       "169  SCN4A                                                 NA\n",
       "\n",
       "[170 rows x 2 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cgc_nodes_10 = extract_node_names(common_gene_compound_answer_10)\n",
    "cgc_nodes_10[['n0','n9']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Controlling the generality and speed with max_connectivity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "ROBOKOP has access to nodes that range from very specific, to highly general. For instance, an anatomical_feature may represent a dopaminergic_neuron, or a neuron, or the midbrain, or the brain or the nervous system.  A biological process may be drawn from anywhere in the very large GO ontology.\n",
    "\n",
    "This has two effects: first, general nodes are highly connected, since e.g. the nervous system is involved in many diseases, where a specfic part of the nervous system may be involved in only very specfic diseases.  Second, because these nodes have many relationships, many paths go through these nodes, and they make the run time of queries much longer.\n",
    "\n",
    "This can be controlled for with the `max_connectivity` parameter, which is the maximum degree of an allowed node in a returned answer.   So if the parameter is set to 1000, then no node with more than 1000 edges in the database will be included in path finding.  The largely has the effect of removing very general nodes, but this is not guaranteed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "two_step_question = make_N_step_question(['disease','biological_process_or_activity','gene'],['MONDO:0019391',None,None])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Return Status: 200\n"
     ]
    }
   ],
   "source": [
    "all_answers = quick(two_step_question,max_results=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "250"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_answers['answers'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Rebuild and caching"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "For details on ROBOKOP caching, and guidance for when to use the `rebuild` parameter, see the notebook on the expand service.  Briefly, the quick service will only run against the cache unless the rebuild parameter is passed, in which case services are re-queried.   The rebuild parameter is passed as part of the json query as seen in the `make_star_query` function above. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Edge Properties"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "In the expand notebook, we saw how a particular predicate could be chosen to limit the results.  This same type information can in fact be added to any edge in a quick query, providing fine control to the results returned."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
